{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup -CHECK SPELLING MISTAKES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import relevant libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libs\n",
    "import os.path\n",
    "from os import path\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import urllib.request\n",
    "from gtfparse import read_gtf\n",
    "from biomart import BiomartServer\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import MinMaxScaler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create function for printing out rows and cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_rowcol(pretext, df):\n",
    "    template = pretext + ': {} rows x {} cols'\n",
    "    print(template.format(str(df.shape[0]), str(df.shape[1])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setup paths - main data path is folder inside the code repository where the data dowloaded from download-data has been stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data'\n",
    "gtf_path = data_path + '/Mus_musculus.GRCm38.99.gtf'\n",
    "tpm_yang_path = data_path + '/GSE90848_Ana6_basal_hair_bulb_TPM.txt'\n",
    "tpm_yang_path2 = data_path + '/GSE90848_Tel_Ana1_Ana2_bulge_HG_basal_HB_TPM.txt'\n",
    "tpm_joost_path = data_path + '/GSE67602_Joost_et_al_expression.txt'\n",
    "tpm_ghahramani_path = data_path + '/GSE99989_NCA_BCatenin_TPM_matrix.csv'\n",
    "gene_tsv_path = data_path + '/gene_names.tsv'\n",
    "gtf_data_path = data_path + '/gtf.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set the main thresholds for the pre-processing\n",
    "- min_num_genes_in_cell_exp - is the min number of genes that need to have expression > **min_ltpm_exp** for it to be considered a valid cell\n",
    "- min_num_cells_for_gene_exp - is the min number of cells that need to have an expression for that gene > **min_ltpm_exp** for it to be considered a valid gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thresholds\n",
    "min_num_genes_in_cell_exp = 1000\n",
    "min_num_cells_for_gene_exp = 500\n",
    "min_ltpm_exp = 1\n",
    "\n",
    "# Train/Test\n",
    "test_data_size = 500"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and process gene data\n",
    "\n",
    "Load GTF data and show some data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (1,24) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded GTF: 1868330 rows x 27 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868325</th>\n",
       "      <td>1868325</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>708</td>\n",
       "      <td>752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868326</th>\n",
       "      <td>1868326</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSMUSE00000997159</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868327</th>\n",
       "      <td>1868327</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868328</th>\n",
       "      <td>1868328</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00001007635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868329</th>\n",
       "      <td>1868329</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1868330 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0     seqname   source     feature    start      end  score  \\\n",
       "0                 0           1   havana        gene  3073253  3074322    NaN   \n",
       "1                 1           1   havana  transcript  3073253  3074322    NaN   \n",
       "2                 2           1   havana        exon  3073253  3074322    NaN   \n",
       "3                 3           1  ensembl        gene  3102016  3102125    NaN   \n",
       "4                 4           1  ensembl  transcript  3102016  3102125    NaN   \n",
       "...             ...         ...      ...         ...      ...      ...    ...   \n",
       "1868325     1868325  JH584295.1  ensembl         CDS      708      752    NaN   \n",
       "1868326     1868326  JH584295.1  ensembl        exon      565      633    NaN   \n",
       "1868327     1868327  JH584295.1  ensembl         CDS      565      633    NaN   \n",
       "1868328     1868328  JH584295.1  ensembl        exon       66      109    NaN   \n",
       "1868329     1868329  JH584295.1  ensembl         CDS       66      109    NaN   \n",
       "\n",
       "        strand  frame             gene_id  ...  transcript_source  \\\n",
       "0            +      0  ENSMUSG00000102693  ...                NaN   \n",
       "1            +      0  ENSMUSG00000102693  ...             havana   \n",
       "2            +      0  ENSMUSG00000102693  ...             havana   \n",
       "3            +      0  ENSMUSG00000064842  ...                NaN   \n",
       "4            +      0  ENSMUSG00000064842  ...            ensembl   \n",
       "...        ...    ...                 ...  ...                ...   \n",
       "1868325      -      2  ENSMUSG00000095742  ...            ensembl   \n",
       "1868326      -      0  ENSMUSG00000095742  ...            ensembl   \n",
       "1868327      -      2  ENSMUSG00000095742  ...            ensembl   \n",
       "1868328      -      0  ENSMUSG00000095742  ...            ensembl   \n",
       "1868329      -      2  ENSMUSG00000095742  ...            ensembl   \n",
       "\n",
       "        transcript_biotype    tag transcript_support_level exon_number  \\\n",
       "0                      NaN    NaN                      NaN         NaN   \n",
       "1                      TEC  basic                      NaN         NaN   \n",
       "2                      TEC  basic                      NaN         1.0   \n",
       "3                      NaN    NaN                      NaN         NaN   \n",
       "4                    snRNA  basic                      NaN         NaN   \n",
       "...                    ...    ...                      ...         ...   \n",
       "1868325     protein_coding  basic                      5.0         5.0   \n",
       "1868326     protein_coding  basic                      5.0         6.0   \n",
       "1868327     protein_coding  basic                      5.0         6.0   \n",
       "1868328     protein_coding  basic                      5.0         7.0   \n",
       "1868329     protein_coding  basic                      5.0         7.0   \n",
       "\n",
       "                    exon_id exon_version ccds_id          protein_id  \\\n",
       "0                       NaN          NaN     NaN                 NaN   \n",
       "1                       NaN          NaN     NaN                 NaN   \n",
       "2        ENSMUSE00001343744          1.0     NaN                 NaN   \n",
       "3                       NaN          NaN     NaN                 NaN   \n",
       "4                       NaN          NaN     NaN                 NaN   \n",
       "...                     ...          ...     ...                 ...   \n",
       "1868325                 NaN          NaN     NaN  ENSMUSP00000137004   \n",
       "1868326  ENSMUSE00000997159          1.0     NaN                 NaN   \n",
       "1868327                 NaN          NaN     NaN  ENSMUSP00000137004   \n",
       "1868328  ENSMUSE00001007635          1.0     NaN                 NaN   \n",
       "1868329                 NaN          NaN     NaN  ENSMUSP00000137004   \n",
       "\n",
       "        protein_version  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  \n",
       "...                 ...  \n",
       "1868325             1.0  \n",
       "1868326             NaN  \n",
       "1868327             1.0  \n",
       "1868328             NaN  \n",
       "1868329             1.0  \n",
       "\n",
       "[1868330 rows x 27 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if path.exists(gtf_data_path) is False:\n",
    "    df_gtf = read_gtf(gtf_path)\n",
    "    df_gtf.to_csv(gtf_data_path)\n",
    "\n",
    "df_gtf = pd.read_csv(gtf_data_path)\n",
    "print_rowcol('Loaded GTF', df_gtf)\n",
    "df_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate feature lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:NumExpr defaulting to 6 threads.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>1069</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>1069</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>1069</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868325</th>\n",
       "      <td>1868325</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>708</td>\n",
       "      <td>752</td>\n",
       "      <td>44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868326</th>\n",
       "      <td>1868326</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSMUSE00000997159</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868327</th>\n",
       "      <td>1868327</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868328</th>\n",
       "      <td>1868328</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00001007635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868329</th>\n",
       "      <td>1868329</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1868330 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0     seqname   source     feature    start      end  \\\n",
       "0                 0           1   havana        gene  3073253  3074322   \n",
       "1                 1           1   havana  transcript  3073253  3074322   \n",
       "2                 2           1   havana        exon  3073253  3074322   \n",
       "3                 3           1  ensembl        gene  3102016  3102125   \n",
       "4                 4           1  ensembl  transcript  3102016  3102125   \n",
       "...             ...         ...      ...         ...      ...      ...   \n",
       "1868325     1868325  JH584295.1  ensembl         CDS      708      752   \n",
       "1868326     1868326  JH584295.1  ensembl        exon      565      633   \n",
       "1868327     1868327  JH584295.1  ensembl         CDS      565      633   \n",
       "1868328     1868328  JH584295.1  ensembl        exon       66      109   \n",
       "1868329     1868329  JH584295.1  ensembl         CDS       66      109   \n",
       "\n",
       "         feature_len  score strand  frame  ... transcript_source  \\\n",
       "0               1069    NaN      +      0  ...               NaN   \n",
       "1               1069    NaN      +      0  ...            havana   \n",
       "2               1069    NaN      +      0  ...            havana   \n",
       "3                109    NaN      +      0  ...               NaN   \n",
       "4                109    NaN      +      0  ...           ensembl   \n",
       "...              ...    ...    ...    ...  ...               ...   \n",
       "1868325           44    NaN      -      2  ...           ensembl   \n",
       "1868326           68    NaN      -      0  ...           ensembl   \n",
       "1868327           68    NaN      -      2  ...           ensembl   \n",
       "1868328           43    NaN      -      0  ...           ensembl   \n",
       "1868329           43    NaN      -      2  ...           ensembl   \n",
       "\n",
       "         transcript_biotype    tag transcript_support_level exon_number  \\\n",
       "0                       NaN    NaN                      NaN         NaN   \n",
       "1                       TEC  basic                      NaN         NaN   \n",
       "2                       TEC  basic                      NaN         1.0   \n",
       "3                       NaN    NaN                      NaN         NaN   \n",
       "4                     snRNA  basic                      NaN         NaN   \n",
       "...                     ...    ...                      ...         ...   \n",
       "1868325      protein_coding  basic                      5.0         5.0   \n",
       "1868326      protein_coding  basic                      5.0         6.0   \n",
       "1868327      protein_coding  basic                      5.0         6.0   \n",
       "1868328      protein_coding  basic                      5.0         7.0   \n",
       "1868329      protein_coding  basic                      5.0         7.0   \n",
       "\n",
       "                    exon_id  exon_version ccds_id          protein_id  \\\n",
       "0                       NaN           NaN     NaN                 NaN   \n",
       "1                       NaN           NaN     NaN                 NaN   \n",
       "2        ENSMUSE00001343744           1.0     NaN                 NaN   \n",
       "3                       NaN           NaN     NaN                 NaN   \n",
       "4                       NaN           NaN     NaN                 NaN   \n",
       "...                     ...           ...     ...                 ...   \n",
       "1868325                 NaN           NaN     NaN  ENSMUSP00000137004   \n",
       "1868326  ENSMUSE00000997159           1.0     NaN                 NaN   \n",
       "1868327                 NaN           NaN     NaN  ENSMUSP00000137004   \n",
       "1868328  ENSMUSE00001007635           1.0     NaN                 NaN   \n",
       "1868329                 NaN           NaN     NaN  ENSMUSP00000137004   \n",
       "\n",
       "        protein_version  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  \n",
       "...                 ...  \n",
       "1868325             1.0  \n",
       "1868326             NaN  \n",
       "1868327             1.0  \n",
       "1868328             NaN  \n",
       "1868329             1.0  \n",
       "\n",
       "[1868330 rows x 28 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf.insert(6,\"feature_len\", df_gtf.end - df_gtf.start)\n",
    "df_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter GTF data for exons, 3' and 5' UTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered GTF for exons and UTR: 1025718 rows x 28 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>1069</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00001343744</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000522066</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213609</td>\n",
       "      <td>3216344</td>\n",
       "      <td>2735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000858910</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3207317</td>\n",
       "      <td>1416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000866652</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3215632</td>\n",
       "      <td>2193</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000863980</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868320</th>\n",
       "      <td>1868320</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>986</td>\n",
       "      <td>1009</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSMUSE00001085113</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868322</th>\n",
       "      <td>1868322</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>972</td>\n",
       "      <td>981</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>ENSMUSE00001070510</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868324</th>\n",
       "      <td>1868324</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>708</td>\n",
       "      <td>752</td>\n",
       "      <td>44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ENSMUSE00001015770</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868326</th>\n",
       "      <td>1868326</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>ENSMUSE00000997159</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868328</th>\n",
       "      <td>1868328</td>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>ENSMUSE00001007635</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1025718 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0     seqname   source feature    start      end  \\\n",
       "2                 2           1   havana    exon  3073253  3074322   \n",
       "5                 5           1  ensembl    exon  3102016  3102125   \n",
       "8                 8           1   havana    exon  3213609  3216344   \n",
       "9                 9           1   havana    exon  3205901  3207317   \n",
       "11               11           1   havana    exon  3213439  3215632   \n",
       "...             ...         ...      ...     ...      ...      ...   \n",
       "1868320     1868320  JH584295.1  ensembl    exon      986     1009   \n",
       "1868322     1868322  JH584295.1  ensembl    exon      972      981   \n",
       "1868324     1868324  JH584295.1  ensembl    exon      708      752   \n",
       "1868326     1868326  JH584295.1  ensembl    exon      565      633   \n",
       "1868328     1868328  JH584295.1  ensembl    exon       66      109   \n",
       "\n",
       "         feature_len  score strand  frame  ... transcript_source  \\\n",
       "2               1069    NaN      +      0  ...            havana   \n",
       "5                109    NaN      +      0  ...           ensembl   \n",
       "8               2735    NaN      -      0  ...            havana   \n",
       "9               1416    NaN      -      0  ...            havana   \n",
       "11              2193    NaN      -      0  ...            havana   \n",
       "...              ...    ...    ...    ...  ...               ...   \n",
       "1868320           23    NaN      -      0  ...           ensembl   \n",
       "1868322            9    NaN      -      0  ...           ensembl   \n",
       "1868324           44    NaN      -      0  ...           ensembl   \n",
       "1868326           68    NaN      -      0  ...           ensembl   \n",
       "1868328           43    NaN      -      0  ...           ensembl   \n",
       "\n",
       "           transcript_biotype    tag transcript_support_level exon_number  \\\n",
       "2                         TEC  basic                      NaN         1.0   \n",
       "5                       snRNA  basic                      NaN         1.0   \n",
       "8        processed_transcript    NaN                      1.0         1.0   \n",
       "9        processed_transcript    NaN                      1.0         2.0   \n",
       "11       processed_transcript    NaN                      1.0         1.0   \n",
       "...                       ...    ...                      ...         ...   \n",
       "1868320        protein_coding  basic                      5.0         3.0   \n",
       "1868322        protein_coding  basic                      5.0         4.0   \n",
       "1868324        protein_coding  basic                      5.0         5.0   \n",
       "1868326        protein_coding  basic                      5.0         6.0   \n",
       "1868328        protein_coding  basic                      5.0         7.0   \n",
       "\n",
       "                    exon_id  exon_version ccds_id protein_id protein_version  \n",
       "2        ENSMUSE00001343744           1.0     NaN        NaN             NaN  \n",
       "5        ENSMUSE00000522066           1.0     NaN        NaN             NaN  \n",
       "8        ENSMUSE00000858910           1.0     NaN        NaN             NaN  \n",
       "9        ENSMUSE00000866652           1.0     NaN        NaN             NaN  \n",
       "11       ENSMUSE00000863980           1.0     NaN        NaN             NaN  \n",
       "...                     ...           ...     ...        ...             ...  \n",
       "1868320  ENSMUSE00001085113           1.0     NaN        NaN             NaN  \n",
       "1868322  ENSMUSE00001070510           1.0     NaN        NaN             NaN  \n",
       "1868324  ENSMUSE00001015770           1.0     NaN        NaN             NaN  \n",
       "1868326  ENSMUSE00000997159           1.0     NaN        NaN             NaN  \n",
       "1868328  ENSMUSE00001007635           1.0     NaN        NaN             NaN  \n",
       "\n",
       "[1025718 rows x 28 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf = df_gtf[(df_gtf.feature=='exon') | (df_gtf.feature=='three_prime_utr') | (df_gtf.feature=='five_prime_utr')]\n",
    "print_rowcol('Filtered GTF for exons and UTR', df_gtf)\n",
    "df_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aggregate the feature lengths to find a transcript length for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>5447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>2240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>7202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>8690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>34434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118636</th>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118637</th>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118638</th>\n",
       "      <td>2398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118639</th>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640</th>\n",
       "      <td>890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55471 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature_len\n",
       "gene_id                        \n",
       "ENSMUSG00000000001         5447\n",
       "ENSMUSG00000000003         2240\n",
       "ENSMUSG00000000028         7202\n",
       "ENSMUSG00000000031         8690\n",
       "ENSMUSG00000000037        34434\n",
       "...                         ...\n",
       "ENSMUSG00000118636          237\n",
       "ENSMUSG00000118637          170\n",
       "ENSMUSG00000118638         2398\n",
       "ENSMUSG00000118639          146\n",
       "ENSMUSG00000118640          890\n",
       "\n",
       "[55471 rows x 1 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf_transcript_len = pd.DataFrame(df_gtf.groupby(['gene_id']).sum()['feature_len'])\n",
    "df_gtf_transcript_len"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check it makes sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>12457</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature_len\n",
       "gene_id                        \n",
       "ENSMUSG00000051951        12457"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf_transcript_len[df_gtf_transcript_len.index.isin(['ENSMUSG00000051951'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213609</td>\n",
       "      <td>3216344</td>\n",
       "      <td>2735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000858910</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3207317</td>\n",
       "      <td>1416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000866652</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3215632</td>\n",
       "      <td>2193</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000863980</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3206523</td>\n",
       "      <td>3207317</td>\n",
       "      <td>794</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000867897</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>946</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSMUSE00000485541</td>\n",
       "      <td>3.0</td>\n",
       "      <td>CCDS14803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSMUSE00000449517</td>\n",
       "      <td>3.0</td>\n",
       "      <td>CCDS14803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216968</td>\n",
       "      <td>2486</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ENSMUSE00000448840</td>\n",
       "      <td>2.0</td>\n",
       "      <td>CCDS14803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>3671349</td>\n",
       "      <td>3671498</td>\n",
       "      <td>149</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS14803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3216021</td>\n",
       "      <td>1539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS14803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0 seqname          source          feature    start      end  \\\n",
       "8            8       1          havana             exon  3213609  3216344   \n",
       "9            9       1          havana             exon  3205901  3207317   \n",
       "11          11       1          havana             exon  3213439  3215632   \n",
       "12          12       1          havana             exon  3206523  3207317   \n",
       "14          14       1  ensembl_havana             exon  3670552  3671498   \n",
       "17          17       1  ensembl_havana             exon  3421702  3421901   \n",
       "19          19       1  ensembl_havana             exon  3214482  3216968   \n",
       "22          22       1  ensembl_havana   five_prime_utr  3671349  3671498   \n",
       "23          23       1  ensembl_havana  three_prime_utr  3214482  3216021   \n",
       "\n",
       "    feature_len  score strand  frame  ... transcript_source  \\\n",
       "8          2735    NaN      -      0  ...            havana   \n",
       "9          1416    NaN      -      0  ...            havana   \n",
       "11         2193    NaN      -      0  ...            havana   \n",
       "12          794    NaN      -      0  ...            havana   \n",
       "14          946    NaN      -      0  ...    ensembl_havana   \n",
       "17          199    NaN      -      0  ...    ensembl_havana   \n",
       "19         2486    NaN      -      0  ...    ensembl_havana   \n",
       "22          149    NaN      -      0  ...    ensembl_havana   \n",
       "23         1539    NaN      -      0  ...    ensembl_havana   \n",
       "\n",
       "      transcript_biotype         tag transcript_support_level exon_number  \\\n",
       "8   processed_transcript         NaN                      1.0         1.0   \n",
       "9   processed_transcript         NaN                      1.0         2.0   \n",
       "11  processed_transcript         NaN                      1.0         1.0   \n",
       "12  processed_transcript         NaN                      1.0         2.0   \n",
       "14        protein_coding  CCDS,basic                      1.0         1.0   \n",
       "17        protein_coding  CCDS,basic                      1.0         2.0   \n",
       "19        protein_coding  CCDS,basic                      1.0         3.0   \n",
       "22        protein_coding  CCDS,basic                      1.0         NaN   \n",
       "23        protein_coding  CCDS,basic                      1.0         NaN   \n",
       "\n",
       "               exon_id  exon_version    ccds_id protein_id protein_version  \n",
       "8   ENSMUSE00000858910           1.0        NaN        NaN             NaN  \n",
       "9   ENSMUSE00000866652           1.0        NaN        NaN             NaN  \n",
       "11  ENSMUSE00000863980           1.0        NaN        NaN             NaN  \n",
       "12  ENSMUSE00000867897           1.0        NaN        NaN             NaN  \n",
       "14  ENSMUSE00000485541           3.0  CCDS14803        NaN             NaN  \n",
       "17  ENSMUSE00000449517           3.0  CCDS14803        NaN             NaN  \n",
       "19  ENSMUSE00000448840           2.0  CCDS14803        NaN             NaN  \n",
       "22                 NaN           NaN  CCDS14803        NaN             NaN  \n",
       "23                 NaN           NaN  CCDS14803        NaN             NaN  \n",
       "\n",
       "[9 rows x 28 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf[df_gtf.gene_id.isin(['ENSMUSG00000051951'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "del df_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert gene lengths to kilobases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>5.447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>2.240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>7.202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>8.690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>34.434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118636</th>\n",
       "      <td>0.237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118637</th>\n",
       "      <td>0.170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118638</th>\n",
       "      <td>2.398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118639</th>\n",
       "      <td>0.146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118640</th>\n",
       "      <td>0.890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55471 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature_len\n",
       "gene_id                        \n",
       "ENSMUSG00000000001        5.447\n",
       "ENSMUSG00000000003        2.240\n",
       "ENSMUSG00000000028        7.202\n",
       "ENSMUSG00000000031        8.690\n",
       "ENSMUSG00000000037       34.434\n",
       "...                         ...\n",
       "ENSMUSG00000118636        0.237\n",
       "ENSMUSG00000118637        0.170\n",
       "ENSMUSG00000118638        2.398\n",
       "ENSMUSG00000118639        0.146\n",
       "ENSMUSG00000118640        0.890\n",
       "\n",
       "[55471 rows x 1 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf_transcript_len = df_gtf_transcript_len / 1000\n",
    "df_gtf_transcript_len"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Resolve Gene names\n",
    "\n",
    "Load gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded gene names: 56289 rows x 1 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064372</th>\n",
       "      <td>mt-Tp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064371</th>\n",
       "      <td>mt-Tt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064370</th>\n",
       "      <td>mt-Cytb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064369</th>\n",
       "      <td>mt-Te</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064368</th>\n",
       "      <td>mt-Nd6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000082803</th>\n",
       "      <td>Gm26460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118095</th>\n",
       "      <td>Gm50435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000106792</th>\n",
       "      <td>Hmgb1-ps4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000107003</th>\n",
       "      <td>2410152P15Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000106796</th>\n",
       "      <td>Gm14968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56289 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_name\n",
       "gene_id                          \n",
       "ENSMUSG00000064372          mt-Tp\n",
       "ENSMUSG00000064371          mt-Tt\n",
       "ENSMUSG00000064370        mt-Cytb\n",
       "ENSMUSG00000064369          mt-Te\n",
       "ENSMUSG00000064368         mt-Nd6\n",
       "...                           ...\n",
       "ENSMUSG00000082803        Gm26460\n",
       "ENSMUSG00000118095        Gm50435\n",
       "ENSMUSG00000106792      Hmgb1-ps4\n",
       "ENSMUSG00000107003  2410152P15Rik\n",
       "ENSMUSG00000106796        Gm14968\n",
       "\n",
       "[56289 rows x 1 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_names = pd.read_csv(gene_tsv_path, sep='\\t', header=None)\n",
    "df_gene_names.columns = ['gene_id', \"gene_name\"]\n",
    "df_gene_names.index = df_gene_names.gene_id\n",
    "df_gene_names = df_gene_names.drop('gene_id', axis=1)\n",
    "print_rowcol('Loaded gene names', df_gene_names)\n",
    "df_gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded gene names: 56289 rows x 1 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mt-Tp</th>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tt</th>\n",
       "      <td>ENSMUSG00000064371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Cytb</th>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Te</th>\n",
       "      <td>ENSMUSG00000064369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm26460</th>\n",
       "      <td>ENSMUSG00000082803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm50435</th>\n",
       "      <td>ENSMUSG00000118095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hmgb1-ps4</th>\n",
       "      <td>ENSMUSG00000106792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410152P15Rik</th>\n",
       "      <td>ENSMUSG00000107003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm14968</th>\n",
       "      <td>ENSMUSG00000106796</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56289 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          gene_id\n",
       "gene_name                        \n",
       "mt-Tp          ENSMUSG00000064372\n",
       "mt-Tt          ENSMUSG00000064371\n",
       "mt-Cytb        ENSMUSG00000064370\n",
       "mt-Te          ENSMUSG00000064369\n",
       "mt-Nd6         ENSMUSG00000064368\n",
       "...                           ...\n",
       "Gm26460        ENSMUSG00000082803\n",
       "Gm50435        ENSMUSG00000118095\n",
       "Hmgb1-ps4      ENSMUSG00000106792\n",
       "2410152P15Rik  ENSMUSG00000107003\n",
       "Gm14968        ENSMUSG00000106796\n",
       "\n",
       "[56289 rows x 1 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_name2id = pd.read_csv(gene_tsv_path, sep='\\t', header=None)\n",
    "df_gene_name2id.columns = ['gene_id', \"gene_name\"]\n",
    "df_gene_name2id.index = df_gene_name2id.gene_name\n",
    "df_gene_name2id = df_gene_name2id.drop('gene_name', axis=1)\n",
    "print_rowcol('Loaded gene names', df_gene_name2id)\n",
    "df_gene_name2id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load in RNA-Seq data\n",
    "\n",
    "### Yang\n",
    "\n",
    "Read the yang1 data set and look at the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Yang1: 46609 rows x 385 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>P1-3-A6</th>\n",
       "      <th>...</th>\n",
       "      <th>P2-1-H11</th>\n",
       "      <th>P2-1-H12</th>\n",
       "      <th>P2-1-H2</th>\n",
       "      <th>P2-1-H3</th>\n",
       "      <th>P2-1-H4</th>\n",
       "      <th>P2-1-H5</th>\n",
       "      <th>P2-1-H6</th>\n",
       "      <th>P2-1-H7</th>\n",
       "      <th>P2-1-H8</th>\n",
       "      <th>P2-1-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001_Gnai3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.60</td>\n",
       "      <td>7.02</td>\n",
       "      <td>17.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>38.31</td>\n",
       "      <td>38.42</td>\n",
       "      <td>47.10</td>\n",
       "      <td>13.19</td>\n",
       "      <td>...</td>\n",
       "      <td>45.27</td>\n",
       "      <td>5.70</td>\n",
       "      <td>13.80</td>\n",
       "      <td>15.89</td>\n",
       "      <td>19.42</td>\n",
       "      <td>3.21</td>\n",
       "      <td>5.65</td>\n",
       "      <td>20.33</td>\n",
       "      <td>16.72</td>\n",
       "      <td>33.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003_Pbsn</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028_Cdc45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.55</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031_H19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>112.13</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037_Scml2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>ERCC-00164</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46605</th>\n",
       "      <td>ERCC-00165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>42.44</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.45</td>\n",
       "      <td>9.12</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>5.41</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>13.15</td>\n",
       "      <td>40.04</td>\n",
       "      <td>9.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>18.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46606</th>\n",
       "      <td>ERCC-00168</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46607</th>\n",
       "      <td>ERCC-00170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46608</th>\n",
       "      <td>ERCC-00171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>143.91</td>\n",
       "      <td>1398.07</td>\n",
       "      <td>328.98</td>\n",
       "      <td>417.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>221.50</td>\n",
       "      <td>695.75</td>\n",
       "      <td>632.11</td>\n",
       "      <td>...</td>\n",
       "      <td>485.62</td>\n",
       "      <td>469.37</td>\n",
       "      <td>290.69</td>\n",
       "      <td>135.58</td>\n",
       "      <td>196.35</td>\n",
       "      <td>388.55</td>\n",
       "      <td>357.51</td>\n",
       "      <td>83.07</td>\n",
       "      <td>220.80</td>\n",
       "      <td>408.54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 385 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_id  P1-3-A1  P1-3-A10  P1-3-A11  P1-3-A12  \\\n",
       "0      ENSMUSG00000000001_Gnai3      0.0     17.60      7.02     17.50   \n",
       "1       ENSMUSG00000000003_Pbsn      0.0      0.00      0.00      0.00   \n",
       "2      ENSMUSG00000000028_Cdc45      0.0      0.00      0.00      0.00   \n",
       "3        ENSMUSG00000000031_H19      0.0      0.00      0.00      0.00   \n",
       "4      ENSMUSG00000000037_Scml2      0.0      0.00      0.00      0.00   \n",
       "...                         ...      ...       ...       ...       ...   \n",
       "46604                ERCC-00164      0.0      0.00      0.00      0.00   \n",
       "46605                ERCC-00165      0.0      0.00     42.44      3.28   \n",
       "46606                ERCC-00168      0.0      0.00      0.00      0.00   \n",
       "46607                ERCC-00170      0.0      0.00      0.00      0.00   \n",
       "46608                ERCC-00171      0.0    143.91   1398.07    328.98   \n",
       "\n",
       "       P1-3-A2  P1-3-A3  P1-3-A4  P1-3-A5  P1-3-A6  ...  P2-1-H11  P2-1-H12  \\\n",
       "0         0.00    38.31    38.42    47.10    13.19  ...     45.27      5.70   \n",
       "1         0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "2         0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "3         0.00     0.00     0.00     0.00   112.13  ...      0.00      6.80   \n",
       "4         0.00     0.00     0.00     0.00     0.00  ...      1.47      0.00   \n",
       "...        ...      ...      ...      ...      ...  ...       ...       ...   \n",
       "46604     0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "46605     0.00     0.00     9.45     9.12     0.00  ...      5.41      1.76   \n",
       "46606     0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "46607     0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "46608   417.57     0.00   221.50   695.75   632.11  ...    485.62    469.37   \n",
       "\n",
       "       P2-1-H2  P2-1-H3  P2-1-H4  P2-1-H5  P2-1-H6  P2-1-H7  P2-1-H8  P2-1-H9  \n",
       "0        13.80    15.89    19.42     3.21     5.65    20.33    16.72    33.57  \n",
       "1         0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "2         0.00     0.00     8.07     0.00     0.00     0.00     2.55     0.00  \n",
       "3         0.00     4.14     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "4         0.00     0.00     0.00     0.00     0.40     0.00     1.32     0.00  \n",
       "...        ...      ...      ...      ...      ...      ...      ...      ...  \n",
       "46604     0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46605     0.00     0.00    13.15    40.04     9.40     0.00    18.05     0.00  \n",
       "46606     0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46607     0.00     2.11     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46608   290.69   135.58   196.35   388.55   357.51    83.07   220.80   408.54  \n",
       "\n",
       "[46609 rows x 385 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_1 = pd.read_csv(tpm_yang_path, sep='\\t')\n",
    "print_rowcol('Loaded Yang1', df_tpm_1)\n",
    "df_tpm_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clean the yang1 dataset to get the gene names, call `.shape` to check we havent filtered anything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filter check: 46609 rows x 3 cols\n"
     ]
    }
   ],
   "source": [
    "# Split gene ids on _ and load into a new data frame and set the columns\n",
    "split_data = pd.DataFrame(df_tpm_1.Gene_id.str.split(\"_\", expand=True))\n",
    "split_data.columns = [\"Gene_id\", \"Gene_name\", \"Gene_name2\"]\n",
    "\n",
    "# Fill in the NA's with blank strings\n",
    "split_data[\"Gene_name\"] = split_data.Gene_name.fillna('')\n",
    "split_data[\"Gene_name2\"] = split_data.Gene_name2.fillna('')\n",
    "\n",
    "# Concatinate the strings that have split more than once back to their standard for e.g GENEID_GENENAME_SOMEMORENAME\n",
    "split_data[\"Gene_name\"] = split_data.apply(lambda x: x.Gene_name if x.Gene_name2 == '' else x.Gene_name + '_' + x.Gene_name2, axis=1)\n",
    "print_rowcol('Filter check', split_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the gene names back into the main dataset, print out dataset to check we do indeed have gene names where they were available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>...</th>\n",
       "      <th>P2-1-H11</th>\n",
       "      <th>P2-1-H12</th>\n",
       "      <th>P2-1-H2</th>\n",
       "      <th>P2-1-H3</th>\n",
       "      <th>P2-1-H4</th>\n",
       "      <th>P2-1-H5</th>\n",
       "      <th>P2-1-H6</th>\n",
       "      <th>P2-1-H7</th>\n",
       "      <th>P2-1-H8</th>\n",
       "      <th>P2-1-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.60</td>\n",
       "      <td>7.02</td>\n",
       "      <td>17.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>38.31</td>\n",
       "      <td>38.42</td>\n",
       "      <td>47.10</td>\n",
       "      <td>...</td>\n",
       "      <td>45.27</td>\n",
       "      <td>5.70</td>\n",
       "      <td>13.80</td>\n",
       "      <td>15.89</td>\n",
       "      <td>19.42</td>\n",
       "      <td>3.21</td>\n",
       "      <td>5.65</td>\n",
       "      <td>20.33</td>\n",
       "      <td>16.72</td>\n",
       "      <td>33.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.55</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031</td>\n",
       "      <td>H19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>ERCC-00164</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46605</th>\n",
       "      <td>ERCC-00165</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>42.44</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.45</td>\n",
       "      <td>9.12</td>\n",
       "      <td>...</td>\n",
       "      <td>5.41</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>13.15</td>\n",
       "      <td>40.04</td>\n",
       "      <td>9.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>18.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46606</th>\n",
       "      <td>ERCC-00168</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46607</th>\n",
       "      <td>ERCC-00170</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46608</th>\n",
       "      <td>ERCC-00171</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>143.91</td>\n",
       "      <td>1398.07</td>\n",
       "      <td>328.98</td>\n",
       "      <td>417.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>221.50</td>\n",
       "      <td>695.75</td>\n",
       "      <td>...</td>\n",
       "      <td>485.62</td>\n",
       "      <td>469.37</td>\n",
       "      <td>290.69</td>\n",
       "      <td>135.58</td>\n",
       "      <td>196.35</td>\n",
       "      <td>388.55</td>\n",
       "      <td>357.51</td>\n",
       "      <td>83.07</td>\n",
       "      <td>220.80</td>\n",
       "      <td>408.54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 386 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene_id Gene_name  P1-3-A1  P1-3-A10  P1-3-A11  P1-3-A12  \\\n",
       "0      ENSMUSG00000000001     Gnai3      0.0     17.60      7.02     17.50   \n",
       "1      ENSMUSG00000000003      Pbsn      0.0      0.00      0.00      0.00   \n",
       "2      ENSMUSG00000000028     Cdc45      0.0      0.00      0.00      0.00   \n",
       "3      ENSMUSG00000000031       H19      0.0      0.00      0.00      0.00   \n",
       "4      ENSMUSG00000000037     Scml2      0.0      0.00      0.00      0.00   \n",
       "...                   ...       ...      ...       ...       ...       ...   \n",
       "46604          ERCC-00164                0.0      0.00      0.00      0.00   \n",
       "46605          ERCC-00165                0.0      0.00     42.44      3.28   \n",
       "46606          ERCC-00168                0.0      0.00      0.00      0.00   \n",
       "46607          ERCC-00170                0.0      0.00      0.00      0.00   \n",
       "46608          ERCC-00171                0.0    143.91   1398.07    328.98   \n",
       "\n",
       "       P1-3-A2  P1-3-A3  P1-3-A4  P1-3-A5  ...  P2-1-H11  P2-1-H12  P2-1-H2  \\\n",
       "0         0.00    38.31    38.42    47.10  ...     45.27      5.70    13.80   \n",
       "1         0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "2         0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "3         0.00     0.00     0.00     0.00  ...      0.00      6.80     0.00   \n",
       "4         0.00     0.00     0.00     0.00  ...      1.47      0.00     0.00   \n",
       "...        ...      ...      ...      ...  ...       ...       ...      ...   \n",
       "46604     0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "46605     0.00     0.00     9.45     9.12  ...      5.41      1.76     0.00   \n",
       "46606     0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "46607     0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "46608   417.57     0.00   221.50   695.75  ...    485.62    469.37   290.69   \n",
       "\n",
       "       P2-1-H3  P2-1-H4  P2-1-H5  P2-1-H6  P2-1-H7  P2-1-H8  P2-1-H9  \n",
       "0        15.89    19.42     3.21     5.65    20.33    16.72    33.57  \n",
       "1         0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "2         0.00     8.07     0.00     0.00     0.00     2.55     0.00  \n",
       "3         4.14     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "4         0.00     0.00     0.00     0.40     0.00     1.32     0.00  \n",
       "...        ...      ...      ...      ...      ...      ...      ...  \n",
       "46604     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46605     0.00    13.15    40.04     9.40     0.00    18.05     0.00  \n",
       "46606     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46607     2.11     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46608   135.58   196.35   388.55   357.51    83.07   220.80   408.54  \n",
       "\n",
       "[46609 rows x 386 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Insert the columns back into the main data array\n",
    "df_tpm_1[\"Gene_id\"] = split_data.Gene_id\n",
    "df_tpm_1.insert(1,\"Gene_name\", split_data.Gene_name)\n",
    "df_tpm_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set gene name to index after checking it is unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_1.Gene_id.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>P1-3-A6</th>\n",
       "      <th>P1-3-A7</th>\n",
       "      <th>...</th>\n",
       "      <th>P2-1-H11</th>\n",
       "      <th>P2-1-H12</th>\n",
       "      <th>P2-1-H2</th>\n",
       "      <th>P2-1-H3</th>\n",
       "      <th>P2-1-H4</th>\n",
       "      <th>P2-1-H5</th>\n",
       "      <th>P2-1-H6</th>\n",
       "      <th>P2-1-H7</th>\n",
       "      <th>P2-1-H8</th>\n",
       "      <th>P2-1-H9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>17.6</td>\n",
       "      <td>7.02</td>\n",
       "      <td>17.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.31</td>\n",
       "      <td>38.42</td>\n",
       "      <td>47.1</td>\n",
       "      <td>13.19</td>\n",
       "      <td>9.66</td>\n",
       "      <td>...</td>\n",
       "      <td>45.27</td>\n",
       "      <td>5.7</td>\n",
       "      <td>13.8</td>\n",
       "      <td>15.89</td>\n",
       "      <td>19.42</td>\n",
       "      <td>3.21</td>\n",
       "      <td>5.65</td>\n",
       "      <td>20.33</td>\n",
       "      <td>16.72</td>\n",
       "      <td>33.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.55</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>112.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 384 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    P1-3-A1  P1-3-A10  P1-3-A11  P1-3-A12  P1-3-A2  P1-3-A3  \\\n",
       "Gene_id                                                                       \n",
       "ENSMUSG00000000001      0.0      17.6      7.02      17.5      0.0    38.31   \n",
       "ENSMUSG00000000003      0.0       0.0      0.00       0.0      0.0     0.00   \n",
       "ENSMUSG00000000028      0.0       0.0      0.00       0.0      0.0     0.00   \n",
       "ENSMUSG00000000031      0.0       0.0      0.00       0.0      0.0     0.00   \n",
       "ENSMUSG00000000037      0.0       0.0      0.00       0.0      0.0     0.00   \n",
       "\n",
       "                    P1-3-A4  P1-3-A5  P1-3-A6  P1-3-A7  ...  P2-1-H11  \\\n",
       "Gene_id                                                 ...             \n",
       "ENSMUSG00000000001    38.42     47.1    13.19     9.66  ...     45.27   \n",
       "ENSMUSG00000000003     0.00      0.0     0.00     0.00  ...      0.00   \n",
       "ENSMUSG00000000028     0.00      0.0     0.00     0.00  ...      0.00   \n",
       "ENSMUSG00000000031     0.00      0.0   112.13     0.00  ...      0.00   \n",
       "ENSMUSG00000000037     0.00      0.0     0.00     0.00  ...      1.47   \n",
       "\n",
       "                    P2-1-H12  P2-1-H2  P2-1-H3  P2-1-H4  P2-1-H5  P2-1-H6  \\\n",
       "Gene_id                                                                     \n",
       "ENSMUSG00000000001       5.7     13.8    15.89    19.42     3.21     5.65   \n",
       "ENSMUSG00000000003       0.0      0.0     0.00     0.00     0.00     0.00   \n",
       "ENSMUSG00000000028       0.0      0.0     0.00     8.07     0.00     0.00   \n",
       "ENSMUSG00000000031       6.8      0.0     4.14     0.00     0.00     0.00   \n",
       "ENSMUSG00000000037       0.0      0.0     0.00     0.00     0.00     0.40   \n",
       "\n",
       "                    P2-1-H7  P2-1-H8  P2-1-H9  \n",
       "Gene_id                                        \n",
       "ENSMUSG00000000001    20.33    16.72    33.57  \n",
       "ENSMUSG00000000003     0.00     0.00     0.00  \n",
       "ENSMUSG00000000028     0.00     2.55     0.00  \n",
       "ENSMUSG00000000031     0.00     0.00     0.00  \n",
       "ENSMUSG00000000037     0.00     1.32     0.00  \n",
       "\n",
       "[5 rows x 384 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_1.index = df_tpm_1['Gene_id']\n",
    "df_tpm_1 = df_tpm_1.drop('Gene_id', axis=1)\n",
    "df_tpm_1 = df_tpm_1.drop('Gene_name', axis=1)\n",
    "df_tpm_1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load in Yang2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Yang2: 46609 rows x 399 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>819b_A1</th>\n",
       "      <th>819b_A10</th>\n",
       "      <th>819b_A11</th>\n",
       "      <th>819b_A2</th>\n",
       "      <th>819b_A3</th>\n",
       "      <th>819b_A4</th>\n",
       "      <th>819b_A5</th>\n",
       "      <th>819b_A6</th>\n",
       "      <th>819b_A7</th>\n",
       "      <th>...</th>\n",
       "      <th>920mat_G7</th>\n",
       "      <th>920mat_G8</th>\n",
       "      <th>920mat_G9</th>\n",
       "      <th>920mat_H11</th>\n",
       "      <th>920mat_H3</th>\n",
       "      <th>920mat_H4</th>\n",
       "      <th>920mat_H5</th>\n",
       "      <th>920mat_H6</th>\n",
       "      <th>920mat_H7</th>\n",
       "      <th>920mat_H8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001_Gnai3</td>\n",
       "      <td>51.73</td>\n",
       "      <td>26.72</td>\n",
       "      <td>29.77</td>\n",
       "      <td>68.41</td>\n",
       "      <td>148.98</td>\n",
       "      <td>55.76</td>\n",
       "      <td>0.61</td>\n",
       "      <td>44.96</td>\n",
       "      <td>29.86</td>\n",
       "      <td>...</td>\n",
       "      <td>20.73</td>\n",
       "      <td>12.29</td>\n",
       "      <td>9.18</td>\n",
       "      <td>14.41</td>\n",
       "      <td>10.34</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.67</td>\n",
       "      <td>28.88</td>\n",
       "      <td>23.65</td>\n",
       "      <td>21.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003_Pbsn</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028_Cdc45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>10.69</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.41</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031_H19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037_Scml2</td>\n",
       "      <td>13.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.73</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.97</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>ERCC-00164</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46605</th>\n",
       "      <td>ERCC-00165</td>\n",
       "      <td>131.21</td>\n",
       "      <td>90.43</td>\n",
       "      <td>13.88</td>\n",
       "      <td>58.42</td>\n",
       "      <td>86.68</td>\n",
       "      <td>75.96</td>\n",
       "      <td>76.02</td>\n",
       "      <td>380.24</td>\n",
       "      <td>9.93</td>\n",
       "      <td>...</td>\n",
       "      <td>72.18</td>\n",
       "      <td>82.65</td>\n",
       "      <td>3.33</td>\n",
       "      <td>106.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>60.24</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>75.92</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46606</th>\n",
       "      <td>ERCC-00168</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46607</th>\n",
       "      <td>ERCC-00170</td>\n",
       "      <td>65.63</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>222.63</td>\n",
       "      <td>222.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>65.14</td>\n",
       "      <td>33.02</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46608</th>\n",
       "      <td>ERCC-00171</td>\n",
       "      <td>3870.71</td>\n",
       "      <td>4852.67</td>\n",
       "      <td>7190.45</td>\n",
       "      <td>5504.13</td>\n",
       "      <td>2273.36</td>\n",
       "      <td>4355.46</td>\n",
       "      <td>2786.26</td>\n",
       "      <td>5251.65</td>\n",
       "      <td>7043.49</td>\n",
       "      <td>...</td>\n",
       "      <td>2908.74</td>\n",
       "      <td>4061.02</td>\n",
       "      <td>1033.23</td>\n",
       "      <td>2030.77</td>\n",
       "      <td>759.40</td>\n",
       "      <td>2250.86</td>\n",
       "      <td>2824.28</td>\n",
       "      <td>1022.18</td>\n",
       "      <td>1960.99</td>\n",
       "      <td>1105.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 399 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_id  819b_A1  819b_A10  819b_A11  819b_A2  \\\n",
       "0      ENSMUSG00000000001_Gnai3    51.73     26.72     29.77    68.41   \n",
       "1       ENSMUSG00000000003_Pbsn     0.00      0.00      0.00     0.00   \n",
       "2      ENSMUSG00000000028_Cdc45     0.00      0.00      0.00     0.00   \n",
       "3        ENSMUSG00000000031_H19     0.00      0.00      0.00     0.00   \n",
       "4      ENSMUSG00000000037_Scml2    13.75      0.00      0.00     4.76   \n",
       "...                         ...      ...       ...       ...      ...   \n",
       "46604                ERCC-00164     0.00      0.00      0.00     0.00   \n",
       "46605                ERCC-00165   131.21     90.43     13.88    58.42   \n",
       "46606                ERCC-00168     0.00      0.00      0.00     0.00   \n",
       "46607                ERCC-00170    65.63      0.00      0.00   222.63   \n",
       "46608                ERCC-00171  3870.71   4852.67   7190.45  5504.13   \n",
       "\n",
       "       819b_A3  819b_A4  819b_A5  819b_A6  819b_A7  ...  920mat_G7  920mat_G8  \\\n",
       "0       148.98    55.76     0.61    44.96    29.86  ...      20.73      12.29   \n",
       "1         0.00     0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "2         0.00     0.00     0.00     0.00     0.00  ...      10.69       0.00   \n",
       "3         0.00     0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "4         0.00     0.00     6.73     0.00     6.97  ...       0.00       0.00   \n",
       "...        ...      ...      ...      ...      ...  ...        ...        ...   \n",
       "46604     0.00     0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "46605    86.68    75.96    76.02   380.24     9.93  ...      72.18      82.65   \n",
       "46606     0.00     0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "46607   222.27     0.00     0.00    65.14    33.02  ...       0.00       0.00   \n",
       "46608  2273.36  4355.46  2786.26  5251.65  7043.49  ...    2908.74    4061.02   \n",
       "\n",
       "       920mat_G9  920mat_H11  920mat_H3  920mat_H4  920mat_H5  920mat_H6  \\\n",
       "0           9.18       14.41      10.34       2.98       0.67      28.88   \n",
       "1           0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "2           0.00        0.00       0.00       9.33       0.00       0.00   \n",
       "3           0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "4           0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "...          ...         ...        ...        ...        ...        ...   \n",
       "46604       0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "46605       3.33      106.09       0.00      60.24       0.00       0.00   \n",
       "46606       0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "46607       0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "46608    1033.23     2030.77     759.40    2250.86    2824.28    1022.18   \n",
       "\n",
       "       920mat_H7  920mat_H8  \n",
       "0          23.65      21.83  \n",
       "1           0.00       0.00  \n",
       "2           1.41       0.00  \n",
       "3           0.00       0.00  \n",
       "4           0.00       0.00  \n",
       "...          ...        ...  \n",
       "46604       0.00       0.00  \n",
       "46605      75.92       0.00  \n",
       "46606       0.00       0.00  \n",
       "46607       0.00       0.00  \n",
       "46608    1960.99    1105.01  \n",
       "\n",
       "[46609 rows x 399 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_2 = pd.read_csv(tpm_yang_path2, sep='\\t')\n",
    "print_rowcol('Loaded Yang2', df_tpm_2)\n",
    "df_tpm_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filter check: 46609 rows x 3 cols\n"
     ]
    }
   ],
   "source": [
    "# Split gene ids on _ and load into a new data frame and set the columns\n",
    "split_data = pd.DataFrame(df_tpm_2.Gene_id.str.split(\"_\", expand=True))\n",
    "split_data.columns = [\"Gene_id\", \"Gene_name\", \"Gene_name2\"]\n",
    "\n",
    "# Fill in the NA's with blank strings\n",
    "split_data[\"Gene_name\"] = split_data.Gene_name.fillna('')\n",
    "split_data[\"Gene_name2\"] = split_data.Gene_name2.fillna('')\n",
    "\n",
    "# Concatinate the strings that have split more than once back to their standard for e.g GENEID_GENENAME_SOMEMORENAME\n",
    "split_data[\"Gene_name\"] = split_data.apply(lambda x: x.Gene_name if x.Gene_name2 == '' else x.Gene_name + '_' + x.Gene_name2, axis=1)\n",
    "print_rowcol('Filter check', split_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>819b_A1</th>\n",
       "      <th>819b_A10</th>\n",
       "      <th>819b_A11</th>\n",
       "      <th>819b_A2</th>\n",
       "      <th>819b_A3</th>\n",
       "      <th>819b_A4</th>\n",
       "      <th>819b_A5</th>\n",
       "      <th>819b_A6</th>\n",
       "      <th>...</th>\n",
       "      <th>920mat_G7</th>\n",
       "      <th>920mat_G8</th>\n",
       "      <th>920mat_G9</th>\n",
       "      <th>920mat_H11</th>\n",
       "      <th>920mat_H3</th>\n",
       "      <th>920mat_H4</th>\n",
       "      <th>920mat_H5</th>\n",
       "      <th>920mat_H6</th>\n",
       "      <th>920mat_H7</th>\n",
       "      <th>920mat_H8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>51.73</td>\n",
       "      <td>26.72</td>\n",
       "      <td>29.77</td>\n",
       "      <td>68.41</td>\n",
       "      <td>148.98</td>\n",
       "      <td>55.76</td>\n",
       "      <td>0.61</td>\n",
       "      <td>44.96</td>\n",
       "      <td>...</td>\n",
       "      <td>20.73</td>\n",
       "      <td>12.29</td>\n",
       "      <td>9.18</td>\n",
       "      <td>14.41</td>\n",
       "      <td>10.34</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.67</td>\n",
       "      <td>28.88</td>\n",
       "      <td>23.65</td>\n",
       "      <td>21.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>10.69</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.41</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031</td>\n",
       "      <td>H19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>13.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.73</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>ERCC-00164</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46605</th>\n",
       "      <td>ERCC-00165</td>\n",
       "      <td></td>\n",
       "      <td>131.21</td>\n",
       "      <td>90.43</td>\n",
       "      <td>13.88</td>\n",
       "      <td>58.42</td>\n",
       "      <td>86.68</td>\n",
       "      <td>75.96</td>\n",
       "      <td>76.02</td>\n",
       "      <td>380.24</td>\n",
       "      <td>...</td>\n",
       "      <td>72.18</td>\n",
       "      <td>82.65</td>\n",
       "      <td>3.33</td>\n",
       "      <td>106.09</td>\n",
       "      <td>0.00</td>\n",
       "      <td>60.24</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>75.92</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46606</th>\n",
       "      <td>ERCC-00168</td>\n",
       "      <td></td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46607</th>\n",
       "      <td>ERCC-00170</td>\n",
       "      <td></td>\n",
       "      <td>65.63</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>222.63</td>\n",
       "      <td>222.27</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>65.14</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46608</th>\n",
       "      <td>ERCC-00171</td>\n",
       "      <td></td>\n",
       "      <td>3870.71</td>\n",
       "      <td>4852.67</td>\n",
       "      <td>7190.45</td>\n",
       "      <td>5504.13</td>\n",
       "      <td>2273.36</td>\n",
       "      <td>4355.46</td>\n",
       "      <td>2786.26</td>\n",
       "      <td>5251.65</td>\n",
       "      <td>...</td>\n",
       "      <td>2908.74</td>\n",
       "      <td>4061.02</td>\n",
       "      <td>1033.23</td>\n",
       "      <td>2030.77</td>\n",
       "      <td>759.40</td>\n",
       "      <td>2250.86</td>\n",
       "      <td>2824.28</td>\n",
       "      <td>1022.18</td>\n",
       "      <td>1960.99</td>\n",
       "      <td>1105.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 400 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene_id Gene_name  819b_A1  819b_A10  819b_A11  819b_A2  \\\n",
       "0      ENSMUSG00000000001     Gnai3    51.73     26.72     29.77    68.41   \n",
       "1      ENSMUSG00000000003      Pbsn     0.00      0.00      0.00     0.00   \n",
       "2      ENSMUSG00000000028     Cdc45     0.00      0.00      0.00     0.00   \n",
       "3      ENSMUSG00000000031       H19     0.00      0.00      0.00     0.00   \n",
       "4      ENSMUSG00000000037     Scml2    13.75      0.00      0.00     4.76   \n",
       "...                   ...       ...      ...       ...       ...      ...   \n",
       "46604          ERCC-00164               0.00      0.00      0.00     0.00   \n",
       "46605          ERCC-00165             131.21     90.43     13.88    58.42   \n",
       "46606          ERCC-00168               0.00      0.00      0.00     0.00   \n",
       "46607          ERCC-00170              65.63      0.00      0.00   222.63   \n",
       "46608          ERCC-00171            3870.71   4852.67   7190.45  5504.13   \n",
       "\n",
       "       819b_A3  819b_A4  819b_A5  819b_A6  ...  920mat_G7  920mat_G8  \\\n",
       "0       148.98    55.76     0.61    44.96  ...      20.73      12.29   \n",
       "1         0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "2         0.00     0.00     0.00     0.00  ...      10.69       0.00   \n",
       "3         0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "4         0.00     0.00     6.73     0.00  ...       0.00       0.00   \n",
       "...        ...      ...      ...      ...  ...        ...        ...   \n",
       "46604     0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "46605    86.68    75.96    76.02   380.24  ...      72.18      82.65   \n",
       "46606     0.00     0.00     0.00     0.00  ...       0.00       0.00   \n",
       "46607   222.27     0.00     0.00    65.14  ...       0.00       0.00   \n",
       "46608  2273.36  4355.46  2786.26  5251.65  ...    2908.74    4061.02   \n",
       "\n",
       "       920mat_G9  920mat_H11  920mat_H3  920mat_H4  920mat_H5  920mat_H6  \\\n",
       "0           9.18       14.41      10.34       2.98       0.67      28.88   \n",
       "1           0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "2           0.00        0.00       0.00       9.33       0.00       0.00   \n",
       "3           0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "4           0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "...          ...         ...        ...        ...        ...        ...   \n",
       "46604       0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "46605       3.33      106.09       0.00      60.24       0.00       0.00   \n",
       "46606       0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "46607       0.00        0.00       0.00       0.00       0.00       0.00   \n",
       "46608    1033.23     2030.77     759.40    2250.86    2824.28    1022.18   \n",
       "\n",
       "       920mat_H7  920mat_H8  \n",
       "0          23.65      21.83  \n",
       "1           0.00       0.00  \n",
       "2           1.41       0.00  \n",
       "3           0.00       0.00  \n",
       "4           0.00       0.00  \n",
       "...          ...        ...  \n",
       "46604       0.00       0.00  \n",
       "46605      75.92       0.00  \n",
       "46606       0.00       0.00  \n",
       "46607       0.00       0.00  \n",
       "46608    1960.99    1105.01  \n",
       "\n",
       "[46609 rows x 400 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Insert the columns back into the main data array\n",
    "df_tpm_2[\"Gene_id\"] = split_data.Gene_id\n",
    "df_tpm_2.insert(1,\"Gene_name\", split_data.Gene_name)\n",
    "df_tpm_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set gene name to index after checking it is unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_2.Gene_id.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>819b_A1</th>\n",
       "      <th>819b_A10</th>\n",
       "      <th>819b_A11</th>\n",
       "      <th>819b_A2</th>\n",
       "      <th>819b_A3</th>\n",
       "      <th>819b_A4</th>\n",
       "      <th>819b_A5</th>\n",
       "      <th>819b_A6</th>\n",
       "      <th>819b_A7</th>\n",
       "      <th>819b_A8</th>\n",
       "      <th>...</th>\n",
       "      <th>920mat_G7</th>\n",
       "      <th>920mat_G8</th>\n",
       "      <th>920mat_G9</th>\n",
       "      <th>920mat_H11</th>\n",
       "      <th>920mat_H3</th>\n",
       "      <th>920mat_H4</th>\n",
       "      <th>920mat_H5</th>\n",
       "      <th>920mat_H6</th>\n",
       "      <th>920mat_H7</th>\n",
       "      <th>920mat_H8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>51.73</td>\n",
       "      <td>26.72</td>\n",
       "      <td>29.77</td>\n",
       "      <td>68.41</td>\n",
       "      <td>148.98</td>\n",
       "      <td>55.76</td>\n",
       "      <td>0.61</td>\n",
       "      <td>44.96</td>\n",
       "      <td>29.86</td>\n",
       "      <td>71.02</td>\n",
       "      <td>...</td>\n",
       "      <td>20.73</td>\n",
       "      <td>12.29</td>\n",
       "      <td>9.18</td>\n",
       "      <td>14.41</td>\n",
       "      <td>10.34</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.67</td>\n",
       "      <td>28.88</td>\n",
       "      <td>23.65</td>\n",
       "      <td>21.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>10.69</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.41</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>13.75</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.73</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 398 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    819b_A1  819b_A10  819b_A11  819b_A2  819b_A3  819b_A4  \\\n",
       "Gene_id                                                                      \n",
       "ENSMUSG00000000001    51.73     26.72     29.77    68.41   148.98    55.76   \n",
       "ENSMUSG00000000003     0.00      0.00      0.00     0.00     0.00     0.00   \n",
       "ENSMUSG00000000028     0.00      0.00      0.00     0.00     0.00     0.00   \n",
       "ENSMUSG00000000031     0.00      0.00      0.00     0.00     0.00     0.00   \n",
       "ENSMUSG00000000037    13.75      0.00      0.00     4.76     0.00     0.00   \n",
       "\n",
       "                    819b_A5  819b_A6  819b_A7  819b_A8  ...  920mat_G7  \\\n",
       "Gene_id                                                 ...              \n",
       "ENSMUSG00000000001     0.61    44.96    29.86    71.02  ...      20.73   \n",
       "ENSMUSG00000000003     0.00     0.00     0.00     0.00  ...       0.00   \n",
       "ENSMUSG00000000028     0.00     0.00     0.00     0.00  ...      10.69   \n",
       "ENSMUSG00000000031     0.00     0.00     0.00     0.00  ...       0.00   \n",
       "ENSMUSG00000000037     6.73     0.00     6.97     0.00  ...       0.00   \n",
       "\n",
       "                    920mat_G8  920mat_G9  920mat_H11  920mat_H3  920mat_H4  \\\n",
       "Gene_id                                                                      \n",
       "ENSMUSG00000000001      12.29       9.18       14.41      10.34       2.98   \n",
       "ENSMUSG00000000003       0.00       0.00        0.00       0.00       0.00   \n",
       "ENSMUSG00000000028       0.00       0.00        0.00       0.00       9.33   \n",
       "ENSMUSG00000000031       0.00       0.00        0.00       0.00       0.00   \n",
       "ENSMUSG00000000037       0.00       0.00        0.00       0.00       0.00   \n",
       "\n",
       "                    920mat_H5  920mat_H6  920mat_H7  920mat_H8  \n",
       "Gene_id                                                         \n",
       "ENSMUSG00000000001       0.67      28.88      23.65      21.83  \n",
       "ENSMUSG00000000003       0.00       0.00       0.00       0.00  \n",
       "ENSMUSG00000000028       0.00       0.00       1.41       0.00  \n",
       "ENSMUSG00000000031       0.00       0.00       0.00       0.00  \n",
       "ENSMUSG00000000037       0.00       0.00       0.00       0.00  \n",
       "\n",
       "[5 rows x 398 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_2.index = df_tpm_2['Gene_id']\n",
    "df_tpm_2 = df_tpm_2.drop('Gene_id', axis=1)\n",
    "df_tpm_2 = df_tpm_2.drop('Gene_name', axis=1)\n",
    "df_tpm_2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Joost / Kasper\n",
    "\n",
    "Load in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Kasper: 26024 rows x 1423 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene\\Cell</th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>363</td>\n",
       "      <td>330</td>\n",
       "      <td>267</td>\n",
       "      <td>262</td>\n",
       "      <td>334</td>\n",
       "      <td>328</td>\n",
       "      <td>280</td>\n",
       "      <td>318</td>\n",
       "      <td>297</td>\n",
       "      <td>...</td>\n",
       "      <td>945</td>\n",
       "      <td>850</td>\n",
       "      <td>696</td>\n",
       "      <td>544</td>\n",
       "      <td>701</td>\n",
       "      <td>817</td>\n",
       "      <td>601</td>\n",
       "      <td>687</td>\n",
       "      <td>589</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>82</td>\n",
       "      <td>79</td>\n",
       "      <td>56</td>\n",
       "      <td>36</td>\n",
       "      <td>54</td>\n",
       "      <td>92</td>\n",
       "      <td>30</td>\n",
       "      <td>64</td>\n",
       "      <td>47</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>276</td>\n",
       "      <td>258</td>\n",
       "      <td>215</td>\n",
       "      <td>190</td>\n",
       "      <td>244</td>\n",
       "      <td>259</td>\n",
       "      <td>209</td>\n",
       "      <td>235</td>\n",
       "      <td>237</td>\n",
       "      <td>...</td>\n",
       "      <td>1000</td>\n",
       "      <td>925</td>\n",
       "      <td>755</td>\n",
       "      <td>667</td>\n",
       "      <td>745</td>\n",
       "      <td>1051</td>\n",
       "      <td>609</td>\n",
       "      <td>749</td>\n",
       "      <td>534</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>25</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>56</td>\n",
       "      <td>49</td>\n",
       "      <td>42</td>\n",
       "      <td>27</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26019</th>\n",
       "      <td>r_MurSAT1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26020</th>\n",
       "      <td>r_HY5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26021</th>\n",
       "      <td>r_tRNA-Arg-CGA_</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26022</th>\n",
       "      <td>r_U14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26023</th>\n",
       "      <td>r_(CGTAG)n</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26024 rows × 1423 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene\\Cell  1772067055_A01  1772067055_A03  1772067055_A04  \\\n",
       "0           ERCC-00002             363             330             267   \n",
       "1           ERCC-00003              20              17              14   \n",
       "2           ERCC-00004             276             258             215   \n",
       "3           ERCC-00009              25              14              14   \n",
       "4           ERCC-00012               0               0               0   \n",
       "...                ...             ...             ...             ...   \n",
       "26019        r_MurSAT1               0               0               0   \n",
       "26020            r_HY5               0               0               0   \n",
       "26021  r_tRNA-Arg-CGA_               0               0               0   \n",
       "26022            r_U14               0               0               0   \n",
       "26023       r_(CGTAG)n               0               0               0   \n",
       "\n",
       "       1772067055_A05  1772067055_A06  1772067055_A07  1772067055_A09  \\\n",
       "0                 262             334             328             280   \n",
       "1                  16              20              26              14   \n",
       "2                 190             244             259             209   \n",
       "3                   7              20              31              20   \n",
       "4                   0               0               0               0   \n",
       "...               ...             ...             ...             ...   \n",
       "26019               1               0               0               0   \n",
       "26020               0               0               0               0   \n",
       "26021               0               0               0               0   \n",
       "26022               0               0               0               0   \n",
       "26023               0               0               0               0   \n",
       "\n",
       "       1772067055_B01  1772067055_B02  ...  1772072285_C12  1772072285_D01  \\\n",
       "0                 318             297  ...             945             850   \n",
       "1                   8              25  ...              82              79   \n",
       "2                 235             237  ...            1000             925   \n",
       "3                  26              24  ...              56              49   \n",
       "4                   0               0  ...               0               0   \n",
       "...               ...             ...  ...             ...             ...   \n",
       "26019               0               0  ...               0               0   \n",
       "26020               0               0  ...               0               0   \n",
       "26021               0               0  ...               0               0   \n",
       "26022               0               0  ...               0               0   \n",
       "26023               0               0  ...               0               0   \n",
       "\n",
       "       1772072285_D03  1772072285_D05  1772072285_D06  1772072285_D09  \\\n",
       "0                 696             544             701             817   \n",
       "1                  56              36              54              92   \n",
       "2                 755             667             745            1051   \n",
       "3                  42              27              33              49   \n",
       "4                   0               0               0               0   \n",
       "...               ...             ...             ...             ...   \n",
       "26019               0               0               0               0   \n",
       "26020               0               0               0               0   \n",
       "26021               0               0               0               0   \n",
       "26022               1               0               0               0   \n",
       "26023               0               0               0               0   \n",
       "\n",
       "       1772072285_D12  1772072285_E02  1772072285_E06  1772072285_G06  \n",
       "0                 601             687             589             229  \n",
       "1                  30              64              47              32  \n",
       "2                 609             749             534             287  \n",
       "3                  30              44              27              13  \n",
       "4                   0               0               0               0  \n",
       "...               ...             ...             ...             ...  \n",
       "26019               0               3               0               0  \n",
       "26020               0               0               0               0  \n",
       "26021               0               0               0               0  \n",
       "26022               0               0               0               0  \n",
       "26023               0               0               0               0  \n",
       "\n",
       "[26024 rows x 1423 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk = pd.read_csv(tpm_joost_path, sep='\\t')\n",
    "print_rowcol('Loaded Kasper', df_rpk)\n",
    "df_rpk"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert from FPKM (Fragments Per Kilobase Million) to TPM (Transcripts Per Kilobase Million)\n",
    "\n",
    "Rename gene/cell column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>363</td>\n",
       "      <td>330</td>\n",
       "      <td>267</td>\n",
       "      <td>262</td>\n",
       "      <td>334</td>\n",
       "      <td>328</td>\n",
       "      <td>280</td>\n",
       "      <td>318</td>\n",
       "      <td>297</td>\n",
       "      <td>...</td>\n",
       "      <td>945</td>\n",
       "      <td>850</td>\n",
       "      <td>696</td>\n",
       "      <td>544</td>\n",
       "      <td>701</td>\n",
       "      <td>817</td>\n",
       "      <td>601</td>\n",
       "      <td>687</td>\n",
       "      <td>589</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>82</td>\n",
       "      <td>79</td>\n",
       "      <td>56</td>\n",
       "      <td>36</td>\n",
       "      <td>54</td>\n",
       "      <td>92</td>\n",
       "      <td>30</td>\n",
       "      <td>64</td>\n",
       "      <td>47</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>276</td>\n",
       "      <td>258</td>\n",
       "      <td>215</td>\n",
       "      <td>190</td>\n",
       "      <td>244</td>\n",
       "      <td>259</td>\n",
       "      <td>209</td>\n",
       "      <td>235</td>\n",
       "      <td>237</td>\n",
       "      <td>...</td>\n",
       "      <td>1000</td>\n",
       "      <td>925</td>\n",
       "      <td>755</td>\n",
       "      <td>667</td>\n",
       "      <td>745</td>\n",
       "      <td>1051</td>\n",
       "      <td>609</td>\n",
       "      <td>749</td>\n",
       "      <td>534</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>25</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>56</td>\n",
       "      <td>49</td>\n",
       "      <td>42</td>\n",
       "      <td>27</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1423 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name  1772067055_A01  1772067055_A03  1772067055_A04  1772067055_A05  \\\n",
       "0  ERCC-00002             363             330             267             262   \n",
       "1  ERCC-00003              20              17              14              16   \n",
       "2  ERCC-00004             276             258             215             190   \n",
       "3  ERCC-00009              25              14              14               7   \n",
       "4  ERCC-00012               0               0               0               0   \n",
       "\n",
       "   1772067055_A06  1772067055_A07  1772067055_A09  1772067055_B01  \\\n",
       "0             334             328             280             318   \n",
       "1              20              26              14               8   \n",
       "2             244             259             209             235   \n",
       "3              20              31              20              26   \n",
       "4               0               0               0               0   \n",
       "\n",
       "   1772067055_B02  ...  1772072285_C12  1772072285_D01  1772072285_D03  \\\n",
       "0             297  ...             945             850             696   \n",
       "1              25  ...              82              79              56   \n",
       "2             237  ...            1000             925             755   \n",
       "3              24  ...              56              49              42   \n",
       "4               0  ...               0               0               0   \n",
       "\n",
       "   1772072285_D05  1772072285_D06  1772072285_D09  1772072285_D12  \\\n",
       "0             544             701             817             601   \n",
       "1              36              54              92              30   \n",
       "2             667             745            1051             609   \n",
       "3              27              33              49              30   \n",
       "4               0               0               0               0   \n",
       "\n",
       "   1772072285_E02  1772072285_E06  1772072285_G06  \n",
       "0             687             589             229  \n",
       "1              64              47              32  \n",
       "2             749             534             287  \n",
       "3              44              27              13  \n",
       "4               0               0               0  \n",
       "\n",
       "[5 rows x 1423 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk.rename( columns={'Gene\\Cell':'gene_name'}, inplace=True)\n",
    "df_rpk.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set gene to index after checking unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk.gene_name.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERCC-00002</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>363</td>\n",
       "      <td>330</td>\n",
       "      <td>267</td>\n",
       "      <td>262</td>\n",
       "      <td>334</td>\n",
       "      <td>328</td>\n",
       "      <td>280</td>\n",
       "      <td>318</td>\n",
       "      <td>297</td>\n",
       "      <td>...</td>\n",
       "      <td>945</td>\n",
       "      <td>850</td>\n",
       "      <td>696</td>\n",
       "      <td>544</td>\n",
       "      <td>701</td>\n",
       "      <td>817</td>\n",
       "      <td>601</td>\n",
       "      <td>687</td>\n",
       "      <td>589</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00003</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>82</td>\n",
       "      <td>79</td>\n",
       "      <td>56</td>\n",
       "      <td>36</td>\n",
       "      <td>54</td>\n",
       "      <td>92</td>\n",
       "      <td>30</td>\n",
       "      <td>64</td>\n",
       "      <td>47</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00004</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>276</td>\n",
       "      <td>258</td>\n",
       "      <td>215</td>\n",
       "      <td>190</td>\n",
       "      <td>244</td>\n",
       "      <td>259</td>\n",
       "      <td>209</td>\n",
       "      <td>235</td>\n",
       "      <td>237</td>\n",
       "      <td>...</td>\n",
       "      <td>1000</td>\n",
       "      <td>925</td>\n",
       "      <td>755</td>\n",
       "      <td>667</td>\n",
       "      <td>745</td>\n",
       "      <td>1051</td>\n",
       "      <td>609</td>\n",
       "      <td>749</td>\n",
       "      <td>534</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00009</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>25</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>26</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>56</td>\n",
       "      <td>49</td>\n",
       "      <td>42</td>\n",
       "      <td>27</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00012</th>\n",
       "      <td>ERCC-00012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1423 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_name  1772067055_A01  1772067055_A03  1772067055_A04  \\\n",
       "gene_name                                                                \n",
       "ERCC-00002  ERCC-00002             363             330             267   \n",
       "ERCC-00003  ERCC-00003              20              17              14   \n",
       "ERCC-00004  ERCC-00004             276             258             215   \n",
       "ERCC-00009  ERCC-00009              25              14              14   \n",
       "ERCC-00012  ERCC-00012               0               0               0   \n",
       "\n",
       "            1772067055_A05  1772067055_A06  1772067055_A07  1772067055_A09  \\\n",
       "gene_name                                                                    \n",
       "ERCC-00002             262             334             328             280   \n",
       "ERCC-00003              16              20              26              14   \n",
       "ERCC-00004             190             244             259             209   \n",
       "ERCC-00009               7              20              31              20   \n",
       "ERCC-00012               0               0               0               0   \n",
       "\n",
       "            1772067055_B01  1772067055_B02  ...  1772072285_C12  \\\n",
       "gene_name                                   ...                   \n",
       "ERCC-00002             318             297  ...             945   \n",
       "ERCC-00003               8              25  ...              82   \n",
       "ERCC-00004             235             237  ...            1000   \n",
       "ERCC-00009              26              24  ...              56   \n",
       "ERCC-00012               0               0  ...               0   \n",
       "\n",
       "            1772072285_D01  1772072285_D03  1772072285_D05  1772072285_D06  \\\n",
       "gene_name                                                                    \n",
       "ERCC-00002             850             696             544             701   \n",
       "ERCC-00003              79              56              36              54   \n",
       "ERCC-00004             925             755             667             745   \n",
       "ERCC-00009              49              42              27              33   \n",
       "ERCC-00012               0               0               0               0   \n",
       "\n",
       "            1772072285_D09  1772072285_D12  1772072285_E02  1772072285_E06  \\\n",
       "gene_name                                                                    \n",
       "ERCC-00002             817             601             687             589   \n",
       "ERCC-00003              92              30              64              47   \n",
       "ERCC-00004            1051             609             749             534   \n",
       "ERCC-00009              49              30              44              27   \n",
       "ERCC-00012               0               0               0               0   \n",
       "\n",
       "            1772072285_G06  \n",
       "gene_name                   \n",
       "ERCC-00002             229  \n",
       "ERCC-00003              32  \n",
       "ERCC-00004             287  \n",
       "ERCC-00009              13  \n",
       "ERCC-00012               0  \n",
       "\n",
       "[5 rows x 1423 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk.index = df_rpk['gene_name']\n",
    "df_rpk.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Insert gene_id into data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape after merging gene ids: 21827 rows x 1422 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>1772067055_B03</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000109644</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000007777</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043644</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000042208</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000020831</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064365</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064371</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064338</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064346</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000064350</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21827 rows × 1422 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    1772067055_A01  1772067055_A03  1772067055_A04  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000109644               0               0               0   \n",
       "ENSMUSG00000007777               1               1               0   \n",
       "ENSMUSG00000043644               0               0               0   \n",
       "ENSMUSG00000042208               0               0               0   \n",
       "ENSMUSG00000020831               0               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000064365               0               0               0   \n",
       "ENSMUSG00000064371               0               0               0   \n",
       "ENSMUSG00000064338               0               0               2   \n",
       "ENSMUSG00000064346               0               2               2   \n",
       "ENSMUSG00000064350               0               0               0   \n",
       "\n",
       "                    1772067055_A05  1772067055_A06  1772067055_A07  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000109644               0               0               0   \n",
       "ENSMUSG00000007777               0               0               0   \n",
       "ENSMUSG00000043644               0               0               0   \n",
       "ENSMUSG00000042208               0               0               0   \n",
       "ENSMUSG00000020831               0               0               1   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000064365               0               0               0   \n",
       "ENSMUSG00000064371               0               0               0   \n",
       "ENSMUSG00000064338               0               1               2   \n",
       "ENSMUSG00000064346               0               0               1   \n",
       "ENSMUSG00000064350               0               1               1   \n",
       "\n",
       "                    1772067055_A09  1772067055_B01  1772067055_B02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000109644               0               0               0   \n",
       "ENSMUSG00000007777               3               0               0   \n",
       "ENSMUSG00000043644               0               0               0   \n",
       "ENSMUSG00000042208               0               0               0   \n",
       "ENSMUSG00000020831               1               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000064365               0               0               0   \n",
       "ENSMUSG00000064371               0               0               1   \n",
       "ENSMUSG00000064338               1               0               3   \n",
       "ENSMUSG00000064346               0               0               0   \n",
       "ENSMUSG00000064350               0               0               0   \n",
       "\n",
       "                    1772067055_B03  ...  1772072285_C12  1772072285_D01  \\\n",
       "gene_id                             ...                                   \n",
       "ENSMUSG00000109644               0  ...               0               0   \n",
       "ENSMUSG00000007777               0  ...               0               0   \n",
       "ENSMUSG00000043644               0  ...               0               0   \n",
       "ENSMUSG00000042208               0  ...               0               0   \n",
       "ENSMUSG00000020831               0  ...               0               0   \n",
       "...                            ...  ...             ...             ...   \n",
       "ENSMUSG00000064365               0  ...               0               0   \n",
       "ENSMUSG00000064371               0  ...               0               1   \n",
       "ENSMUSG00000064338               0  ...               3               3   \n",
       "ENSMUSG00000064346               2  ...               3               1   \n",
       "ENSMUSG00000064350               0  ...               2               1   \n",
       "\n",
       "                    1772072285_D03  1772072285_D05  1772072285_D06  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000109644               0               0               0   \n",
       "ENSMUSG00000007777               0               0               1   \n",
       "ENSMUSG00000043644               0               0               0   \n",
       "ENSMUSG00000042208               0               0               0   \n",
       "ENSMUSG00000020831               0               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000064365               0               0               0   \n",
       "ENSMUSG00000064371               0               0               0   \n",
       "ENSMUSG00000064338               0               1               1   \n",
       "ENSMUSG00000064346               0               3               0   \n",
       "ENSMUSG00000064350               0               0               1   \n",
       "\n",
       "                    1772072285_D09  1772072285_D12  1772072285_E02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000109644               0               0               0   \n",
       "ENSMUSG00000007777               0               0               0   \n",
       "ENSMUSG00000043644               0               0               0   \n",
       "ENSMUSG00000042208               0               0               0   \n",
       "ENSMUSG00000020831               0               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000064365               0               0               0   \n",
       "ENSMUSG00000064371               0               0               0   \n",
       "ENSMUSG00000064338               5               0               3   \n",
       "ENSMUSG00000064346               0               1               0   \n",
       "ENSMUSG00000064350               0               1               0   \n",
       "\n",
       "                    1772072285_E06  1772072285_G06  \n",
       "gene_id                                             \n",
       "ENSMUSG00000109644               0               0  \n",
       "ENSMUSG00000007777               0               2  \n",
       "ENSMUSG00000043644               0               0  \n",
       "ENSMUSG00000042208               0               0  \n",
       "ENSMUSG00000020831               0               0  \n",
       "...                            ...             ...  \n",
       "ENSMUSG00000064365               0               0  \n",
       "ENSMUSG00000064371               1               0  \n",
       "ENSMUSG00000064338               1               0  \n",
       "ENSMUSG00000064346               2               0  \n",
       "ENSMUSG00000064350               4               0  \n",
       "\n",
       "[21827 rows x 1422 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk_merged = df_rpk\n",
    "df_rpk_merged = df_rpk_merged.drop('gene_name', axis=1)\n",
    "df_rpk_merged = df_gene_name2id.join(df_rpk_merged, lsuffix='', rsuffix='', how='inner')\n",
    "df_rpk_merged.index = df_rpk_merged.gene_id\n",
    "df_rpk_merged = df_rpk_merged.drop('gene_id', axis=1)\n",
    "print_rowcol('Shape after merging gene ids', df_rpk_merged)\n",
    "df_rpk_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get length of each gene into data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape after merging gene feature_len: 21522 rows x 1423 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_len</th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>5.447</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>2.240</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>7.202</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>8.690</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>34.434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118401</th>\n",
       "      <td>2.185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118406</th>\n",
       "      <td>0.396</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118491</th>\n",
       "      <td>4.836</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118506</th>\n",
       "      <td>6.805</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118560</th>\n",
       "      <td>1.238</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21522 rows × 1423 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature_len  1772067055_A01  1772067055_A03  \\\n",
       "gene_id                                                           \n",
       "ENSMUSG00000000001        5.447               1               0   \n",
       "ENSMUSG00000000003        2.240               0               0   \n",
       "ENSMUSG00000000028        7.202               0               0   \n",
       "ENSMUSG00000000031        8.690               0               0   \n",
       "ENSMUSG00000000037       34.434               0               0   \n",
       "...                         ...             ...             ...   \n",
       "ENSMUSG00000118401        2.185               0               0   \n",
       "ENSMUSG00000118406        0.396               0               0   \n",
       "ENSMUSG00000118491        4.836               0               0   \n",
       "ENSMUSG00000118506        6.805               0               0   \n",
       "ENSMUSG00000118560        1.238               0               0   \n",
       "\n",
       "                    1772067055_A04  1772067055_A05  1772067055_A06  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001               1               0               0   \n",
       "ENSMUSG00000000003               0               0               0   \n",
       "ENSMUSG00000000028               0               0               0   \n",
       "ENSMUSG00000000031               0               0               0   \n",
       "ENSMUSG00000000037               0               1               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401               0               0               0   \n",
       "ENSMUSG00000118406               0               0               0   \n",
       "ENSMUSG00000118491               0               2               0   \n",
       "ENSMUSG00000118506               0               0               0   \n",
       "ENSMUSG00000118560               0               0               0   \n",
       "\n",
       "                    1772067055_A07  1772067055_A09  1772067055_B01  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001               0               1               0   \n",
       "ENSMUSG00000000003               0               0               0   \n",
       "ENSMUSG00000000028               0               0               3   \n",
       "ENSMUSG00000000031               0               0               0   \n",
       "ENSMUSG00000000037               0               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401               0               0               0   \n",
       "ENSMUSG00000118406               0               0               0   \n",
       "ENSMUSG00000118491               2               0               0   \n",
       "ENSMUSG00000118506               0               0               0   \n",
       "ENSMUSG00000118560               0               0               0   \n",
       "\n",
       "                    1772067055_B02  ...  1772072285_C12  1772072285_D01  \\\n",
       "gene_id                             ...                                   \n",
       "ENSMUSG00000000001               0  ...               4               1   \n",
       "ENSMUSG00000000003               0  ...               0               0   \n",
       "ENSMUSG00000000028               0  ...               0               0   \n",
       "ENSMUSG00000000031               0  ...               0               0   \n",
       "ENSMUSG00000000037               0  ...               0               0   \n",
       "...                            ...  ...             ...             ...   \n",
       "ENSMUSG00000118401               0  ...               0               0   \n",
       "ENSMUSG00000118406               0  ...               0               0   \n",
       "ENSMUSG00000118491               0  ...               2               2   \n",
       "ENSMUSG00000118506               0  ...               0               0   \n",
       "ENSMUSG00000118560               0  ...               0               0   \n",
       "\n",
       "                    1772072285_D03  1772072285_D05  1772072285_D06  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001               3               0               0   \n",
       "ENSMUSG00000000003               0               0               0   \n",
       "ENSMUSG00000000028               0               0               0   \n",
       "ENSMUSG00000000031               0               0               0   \n",
       "ENSMUSG00000000037               0               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401               0               0               0   \n",
       "ENSMUSG00000118406               0               0               0   \n",
       "ENSMUSG00000118491               3               0               0   \n",
       "ENSMUSG00000118506               0               0               0   \n",
       "ENSMUSG00000118560               0               0               0   \n",
       "\n",
       "                    1772072285_D09  1772072285_D12  1772072285_E02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001               1               0               2   \n",
       "ENSMUSG00000000003               0               0               0   \n",
       "ENSMUSG00000000028               0               0               0   \n",
       "ENSMUSG00000000031               0               0               0   \n",
       "ENSMUSG00000000037               0               0               0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401               0               0               0   \n",
       "ENSMUSG00000118406               0               0               0   \n",
       "ENSMUSG00000118491               1               0               1   \n",
       "ENSMUSG00000118506               0               0               0   \n",
       "ENSMUSG00000118560               0               0               0   \n",
       "\n",
       "                    1772072285_E06  1772072285_G06  \n",
       "gene_id                                             \n",
       "ENSMUSG00000000001               0               0  \n",
       "ENSMUSG00000000003               0               0  \n",
       "ENSMUSG00000000028               0               0  \n",
       "ENSMUSG00000000031               0               0  \n",
       "ENSMUSG00000000037               0               0  \n",
       "...                            ...             ...  \n",
       "ENSMUSG00000118401               0               0  \n",
       "ENSMUSG00000118406               0               0  \n",
       "ENSMUSG00000118491               0               1  \n",
       "ENSMUSG00000118506               0               0  \n",
       "ENSMUSG00000118560               0               0  \n",
       "\n",
       "[21522 rows x 1423 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk_merged_len = df_gtf_transcript_len.join(df_rpk_merged, lsuffix='', rsuffix='', how='inner')\n",
    "print_rowcol('Shape after merging gene feature_len', df_rpk_merged_len)\n",
    "df_rpk_merged_len"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sum the read counts per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaling_factor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1772067055_A01</th>\n",
       "      <td>0.004375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772067055_A03</th>\n",
       "      <td>0.004793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772067055_A04</th>\n",
       "      <td>0.005413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772067055_A05</th>\n",
       "      <td>0.002196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772067055_A06</th>\n",
       "      <td>0.003068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772072285_D09</th>\n",
       "      <td>0.007320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772072285_D12</th>\n",
       "      <td>0.003783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772072285_E02</th>\n",
       "      <td>0.011444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772072285_E06</th>\n",
       "      <td>0.003907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772072285_G06</th>\n",
       "      <td>0.002564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1422 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                scaling_factor\n",
       "1772067055_A01        0.004375\n",
       "1772067055_A03        0.004793\n",
       "1772067055_A04        0.005413\n",
       "1772067055_A05        0.002196\n",
       "1772067055_A06        0.003068\n",
       "...                        ...\n",
       "1772072285_D09        0.007320\n",
       "1772072285_D12        0.003783\n",
       "1772072285_E02        0.011444\n",
       "1772072285_E06        0.003907\n",
       "1772072285_G06        0.002564\n",
       "\n",
       "[1422 rows x 1 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_scaling_factor = pd.DataFrame(df_rpk_merged_len.sum(axis=0) / 1000000)\n",
    "df_scaling_factor.columns = ['scaling_factor']\n",
    "df_scaling_factor = df_scaling_factor.drop('feature_len')\n",
    "df_scaling_factor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Divide the read counts by the length of each gene in kilobases. This gives you reads per kilobase (RPK)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>1772067055_B03</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.183587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.183587</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.183587</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.734349</td>\n",
       "      <td>0.183587</td>\n",
       "      <td>0.550762</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.183587</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.367175</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.416551</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029041</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118401</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118406</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118491</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.413565</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.413565</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.413565</td>\n",
       "      <td>0.413565</td>\n",
       "      <td>0.620347</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.206782</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.206782</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.206782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118506</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118560</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21522 rows × 1422 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    1772067055_A01  1772067055_A03  1772067055_A04  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        0.183587             0.0        0.183587   \n",
       "ENSMUSG00000000003        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000028        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000031        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000037        0.000000             0.0        0.000000   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118406        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118491        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118506        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118560        0.000000             0.0        0.000000   \n",
       "\n",
       "                    1772067055_A05  1772067055_A06  1772067055_A07  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000003        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000028        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000031        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000037        0.029041             0.0        0.000000   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118406        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118491        0.413565             0.0        0.413565   \n",
       "ENSMUSG00000118506        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118560        0.000000             0.0        0.000000   \n",
       "\n",
       "                    1772067055_A09  1772067055_B01  1772067055_B02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        0.183587        0.000000             0.0   \n",
       "ENSMUSG00000000003        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000000028        0.000000        0.416551             0.0   \n",
       "ENSMUSG00000000031        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000000037        0.000000        0.000000             0.0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118406        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118491        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118506        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118560        0.000000        0.000000             0.0   \n",
       "\n",
       "                    1772067055_B03  ...  1772072285_C12  1772072285_D01  \\\n",
       "gene_id                             ...                                   \n",
       "ENSMUSG00000000001             0.0  ...        0.734349        0.183587   \n",
       "ENSMUSG00000000003             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000000028             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000000031             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000000037             0.0  ...        0.000000        0.000000   \n",
       "...                            ...  ...             ...             ...   \n",
       "ENSMUSG00000118401             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000118406             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000118491             0.0  ...        0.413565        0.413565   \n",
       "ENSMUSG00000118506             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000118560             0.0  ...        0.000000        0.000000   \n",
       "\n",
       "                    1772072285_D03  1772072285_D05  1772072285_D06  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        0.550762             0.0             0.0   \n",
       "ENSMUSG00000000003        0.000000             0.0             0.0   \n",
       "ENSMUSG00000000028        0.000000             0.0             0.0   \n",
       "ENSMUSG00000000031        0.000000             0.0             0.0   \n",
       "ENSMUSG00000000037        0.000000             0.0             0.0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0             0.0   \n",
       "ENSMUSG00000118406        0.000000             0.0             0.0   \n",
       "ENSMUSG00000118491        0.620347             0.0             0.0   \n",
       "ENSMUSG00000118506        0.000000             0.0             0.0   \n",
       "ENSMUSG00000118560        0.000000             0.0             0.0   \n",
       "\n",
       "                    1772072285_D09  1772072285_D12  1772072285_E02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        0.183587             0.0        0.367175   \n",
       "ENSMUSG00000000003        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000028        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000031        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000037        0.000000             0.0        0.000000   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118406        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118491        0.206782             0.0        0.206782   \n",
       "ENSMUSG00000118506        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118560        0.000000             0.0        0.000000   \n",
       "\n",
       "                    1772072285_E06  1772072285_G06  \n",
       "gene_id                                             \n",
       "ENSMUSG00000000001             0.0        0.000000  \n",
       "ENSMUSG00000000003             0.0        0.000000  \n",
       "ENSMUSG00000000028             0.0        0.000000  \n",
       "ENSMUSG00000000031             0.0        0.000000  \n",
       "ENSMUSG00000000037             0.0        0.000000  \n",
       "...                            ...             ...  \n",
       "ENSMUSG00000118401             0.0        0.000000  \n",
       "ENSMUSG00000118406             0.0        0.000000  \n",
       "ENSMUSG00000118491             0.0        0.206782  \n",
       "ENSMUSG00000118506             0.0        0.000000  \n",
       "ENSMUSG00000118560             0.0        0.000000  \n",
       "\n",
       "[21522 rows x 1422 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_rpk_merged_len = df_rpk_merged_len.iloc[:,1:].div(df_rpk_merged_len.feature_len, axis=0)\n",
    "df_rpk_merged_len"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Divide the RPK values by the “per million” scaling factor. This gives you TPM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1772067055_A01</th>\n",
       "      <th>1772067055_A03</th>\n",
       "      <th>1772067055_A04</th>\n",
       "      <th>1772067055_A05</th>\n",
       "      <th>1772067055_A06</th>\n",
       "      <th>1772067055_A07</th>\n",
       "      <th>1772067055_A09</th>\n",
       "      <th>1772067055_B01</th>\n",
       "      <th>1772067055_B02</th>\n",
       "      <th>1772067055_B03</th>\n",
       "      <th>...</th>\n",
       "      <th>1772072285_C12</th>\n",
       "      <th>1772072285_D01</th>\n",
       "      <th>1772072285_D03</th>\n",
       "      <th>1772072285_D05</th>\n",
       "      <th>1772072285_D06</th>\n",
       "      <th>1772072285_D09</th>\n",
       "      <th>1772072285_D12</th>\n",
       "      <th>1772072285_E02</th>\n",
       "      <th>1772072285_E06</th>\n",
       "      <th>1772072285_G06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>41.96281</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.915998</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21.552864</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>80.099169</td>\n",
       "      <td>31.137601</td>\n",
       "      <td>127.520696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.080232</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.084463</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>93.649046</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.224528</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118401</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118406</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118491</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>188.326471</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.369229</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>45.109613</td>\n",
       "      <td>70.143306</td>\n",
       "      <td>143.632182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.248971</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.069072</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80.648387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118506</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000118560</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21522 rows × 1422 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    1772067055_A01  1772067055_A03  1772067055_A04  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        41.96281             0.0       33.915998   \n",
       "ENSMUSG00000000003         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000000028         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000000031         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000000037         0.00000             0.0        0.000000   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000118406         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000118491         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000118506         0.00000             0.0        0.000000   \n",
       "ENSMUSG00000118560         0.00000             0.0        0.000000   \n",
       "\n",
       "                    1772067055_A05  1772067055_A06  1772067055_A07  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000003        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000028        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000031        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000037       13.224528             0.0        0.000000   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118406        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118491      188.326471             0.0       34.369229   \n",
       "ENSMUSG00000118506        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118560        0.000000             0.0        0.000000   \n",
       "\n",
       "                    1772067055_A09  1772067055_B01  1772067055_B02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001       21.552864        0.000000             0.0   \n",
       "ENSMUSG00000000003        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000000028        0.000000       93.649046             0.0   \n",
       "ENSMUSG00000000031        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000000037        0.000000        0.000000             0.0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118406        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118491        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118506        0.000000        0.000000             0.0   \n",
       "ENSMUSG00000118560        0.000000        0.000000             0.0   \n",
       "\n",
       "                    1772067055_B03  ...  1772072285_C12  1772072285_D01  \\\n",
       "gene_id                             ...                                   \n",
       "ENSMUSG00000000001             0.0  ...       80.099169       31.137601   \n",
       "ENSMUSG00000000003             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000000028             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000000031             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000000037             0.0  ...        0.000000        0.000000   \n",
       "...                            ...  ...             ...             ...   \n",
       "ENSMUSG00000118401             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000118406             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000118491             0.0  ...       45.109613       70.143306   \n",
       "ENSMUSG00000118506             0.0  ...        0.000000        0.000000   \n",
       "ENSMUSG00000118560             0.0  ...        0.000000        0.000000   \n",
       "\n",
       "                    1772072285_D03  1772072285_D05  1772072285_D06  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001      127.520696             0.0             0.0   \n",
       "ENSMUSG00000000003        0.000000             0.0             0.0   \n",
       "ENSMUSG00000000028        0.000000             0.0             0.0   \n",
       "ENSMUSG00000000031        0.000000             0.0             0.0   \n",
       "ENSMUSG00000000037        0.000000             0.0             0.0   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0             0.0   \n",
       "ENSMUSG00000118406        0.000000             0.0             0.0   \n",
       "ENSMUSG00000118491      143.632182             0.0             0.0   \n",
       "ENSMUSG00000118506        0.000000             0.0             0.0   \n",
       "ENSMUSG00000118560        0.000000             0.0             0.0   \n",
       "\n",
       "                    1772072285_D09  1772072285_D12  1772072285_E02  \\\n",
       "gene_id                                                              \n",
       "ENSMUSG00000000001       25.080232             0.0       32.084463   \n",
       "ENSMUSG00000000003        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000028        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000031        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000000037        0.000000             0.0        0.000000   \n",
       "...                            ...             ...             ...   \n",
       "ENSMUSG00000118401        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118406        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118491       28.248971             0.0       18.069072   \n",
       "ENSMUSG00000118506        0.000000             0.0        0.000000   \n",
       "ENSMUSG00000118560        0.000000             0.0        0.000000   \n",
       "\n",
       "                    1772072285_E06  1772072285_G06  \n",
       "gene_id                                             \n",
       "ENSMUSG00000000001             0.0        0.000000  \n",
       "ENSMUSG00000000003             0.0        0.000000  \n",
       "ENSMUSG00000000028             0.0        0.000000  \n",
       "ENSMUSG00000000031             0.0        0.000000  \n",
       "ENSMUSG00000000037             0.0        0.000000  \n",
       "...                            ...             ...  \n",
       "ENSMUSG00000118401             0.0        0.000000  \n",
       "ENSMUSG00000118406             0.0        0.000000  \n",
       "ENSMUSG00000118491             0.0       80.648387  \n",
       "ENSMUSG00000118506             0.0        0.000000  \n",
       "ENSMUSG00000118560             0.0        0.000000  \n",
       "\n",
       "[21522 rows x 1422 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_3 = df_rpk_merged_len.div(df_scaling_factor.scaling_factor, axis=1)\n",
    "df_tpm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ghahramani"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load in dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded Ghahramani: 14145 rows x 295 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>AG1neg_1</th>\n",
       "      <th>AG2pos_2</th>\n",
       "      <th>AG1neg_3</th>\n",
       "      <th>AG2pos_4</th>\n",
       "      <th>AG1neg_5</th>\n",
       "      <th>AG2pos_6</th>\n",
       "      <th>AG1neg_7</th>\n",
       "      <th>AG2pos_8</th>\n",
       "      <th>AG1neg_9</th>\n",
       "      <th>...</th>\n",
       "      <th>GHA501A86</th>\n",
       "      <th>GHA501A88</th>\n",
       "      <th>GHA501A89</th>\n",
       "      <th>GHA501A90</th>\n",
       "      <th>GHA501A93</th>\n",
       "      <th>GHA501A94</th>\n",
       "      <th>GHA501A95</th>\n",
       "      <th>GHA501A98</th>\n",
       "      <th>GHA501A99</th>\n",
       "      <th>GHA501A9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>122.014549</td>\n",
       "      <td>7.484987</td>\n",
       "      <td>597.717004</td>\n",
       "      <td>643.657745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>331.238736</td>\n",
       "      <td>117.133604</td>\n",
       "      <td>468.887523</td>\n",
       "      <td>243.073894</td>\n",
       "      <td>...</td>\n",
       "      <td>61.411894</td>\n",
       "      <td>230.333860</td>\n",
       "      <td>98.040207</td>\n",
       "      <td>93.716396</td>\n",
       "      <td>673.177033</td>\n",
       "      <td>50.220174</td>\n",
       "      <td>303.137672</td>\n",
       "      <td>85.992001</td>\n",
       "      <td>14.915042</td>\n",
       "      <td>4.671418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.560498</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.958870</td>\n",
       "      <td>2.201127</td>\n",
       "      <td>1.425608</td>\n",
       "      <td>222.350975</td>\n",
       "      <td>10.621630</td>\n",
       "      <td>433.142296</td>\n",
       "      <td>1.590745</td>\n",
       "      <td>540.554230</td>\n",
       "      <td>2.841134</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Apoh</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.569181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>174.856508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Narf</td>\n",
       "      <td>1.475207</td>\n",
       "      <td>7.818909</td>\n",
       "      <td>111.868532</td>\n",
       "      <td>117.207838</td>\n",
       "      <td>8.548119</td>\n",
       "      <td>34.838241</td>\n",
       "      <td>49.003077</td>\n",
       "      <td>50.385245</td>\n",
       "      <td>107.228871</td>\n",
       "      <td>...</td>\n",
       "      <td>7.230203</td>\n",
       "      <td>2.514732</td>\n",
       "      <td>1.954466</td>\n",
       "      <td>2.185207</td>\n",
       "      <td>8.604779</td>\n",
       "      <td>3.643098</td>\n",
       "      <td>219.721960</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.016629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cav2</td>\n",
       "      <td>160.239035</td>\n",
       "      <td>278.810852</td>\n",
       "      <td>112.274304</td>\n",
       "      <td>41.820736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28.127109</td>\n",
       "      <td>108.345577</td>\n",
       "      <td>65.333237</td>\n",
       "      <td>4.705999</td>\n",
       "      <td>...</td>\n",
       "      <td>2.163511</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40.207721</td>\n",
       "      <td>1.198789</td>\n",
       "      <td>241.109633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.598203</td>\n",
       "      <td>221.756227</td>\n",
       "      <td>59.831210</td>\n",
       "      <td>7.808018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14140</th>\n",
       "      <td>Exosc6</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37.295445</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14141</th>\n",
       "      <td>Pde2a</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14142</th>\n",
       "      <td>Flt3l</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.609519</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18.984288</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.020048</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14143</th>\n",
       "      <td>Rnf223</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.030893</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14144</th>\n",
       "      <td>Cmtm1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>54.027996</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14145 rows × 295 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0    AG1neg_1    AG2pos_2    AG1neg_3    AG2pos_4   AG1neg_5  \\\n",
       "0          Gnai3  122.014549    7.484987  597.717004  643.657745   0.000000   \n",
       "1          Cdc45    0.000000    0.000000    0.000000    5.560498   0.000000   \n",
       "2           Apoh    0.000000    0.000000    0.000000    0.000000   0.000000   \n",
       "3           Narf    1.475207    7.818909  111.868532  117.207838   8.548119   \n",
       "4           Cav2  160.239035  278.810852  112.274304   41.820736   0.000000   \n",
       "...          ...         ...         ...         ...         ...        ...   \n",
       "14140     Exosc6    0.000000    0.000000    0.000000    0.000000   0.000000   \n",
       "14141      Pde2a    0.000000    0.000000    0.000000    0.000000   0.000000   \n",
       "14142      Flt3l    0.000000    0.609519    0.000000    0.000000   0.000000   \n",
       "14143     Rnf223    0.000000    0.000000    0.000000   11.030893   0.000000   \n",
       "14144      Cmtm1    0.000000    0.000000    0.000000    0.000000  54.027996   \n",
       "\n",
       "         AG2pos_6    AG1neg_7    AG2pos_8    AG1neg_9  ...  GHA501A86  \\\n",
       "0      331.238736  117.133604  468.887523  243.073894  ...  61.411894   \n",
       "1        0.000000    0.000000    0.000000    0.000000  ...   0.958870   \n",
       "2        0.000000    0.000000    0.000000    0.000000  ...   0.000000   \n",
       "3       34.838241   49.003077   50.385245  107.228871  ...   7.230203   \n",
       "4       28.127109  108.345577   65.333237    4.705999  ...   2.163511   \n",
       "...           ...         ...         ...         ...  ...        ...   \n",
       "14140    0.000000   37.295445    0.000000    0.000000  ...   0.000000   \n",
       "14141    0.000000    0.000000    0.000000    0.000000  ...   0.000000   \n",
       "14142    0.000000   18.984288    0.000000    0.000000  ...   0.000000   \n",
       "14143    0.000000    0.000000    0.000000    0.000000  ...   0.000000   \n",
       "14144    0.000000    0.000000    0.000000    0.000000  ...   0.000000   \n",
       "\n",
       "        GHA501A88  GHA501A89   GHA501A90   GHA501A93   GHA501A94   GHA501A95  \\\n",
       "0      230.333860  98.040207   93.716396  673.177033   50.220174  303.137672   \n",
       "1        2.201127   1.425608  222.350975   10.621630  433.142296    1.590745   \n",
       "2        0.000000   0.000000    0.000000    0.000000    0.000000    2.569181   \n",
       "3        2.514732   1.954466    2.185207    8.604779    3.643098  219.721960   \n",
       "4        0.000000  40.207721    1.198789  241.109633    0.000000    0.598203   \n",
       "...           ...        ...         ...         ...         ...         ...   \n",
       "14140    0.000000   0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "14141    0.000000   0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "14142    0.000000   0.000000    0.000000    0.000000    0.000000    1.020048   \n",
       "14143    0.000000   0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "14144    0.000000   0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "        GHA501A98  GHA501A99    GHA501A9  \n",
       "0       85.992001  14.915042    4.671418  \n",
       "1      540.554230   2.841134    0.000000  \n",
       "2        0.000000   0.000000  174.856508  \n",
       "3        0.000000   0.000000    1.016629  \n",
       "4      221.756227  59.831210    7.808018  \n",
       "...           ...        ...         ...  \n",
       "14140    0.000000   0.000000    0.000000  \n",
       "14141    0.000000   0.000000    0.000000  \n",
       "14142    0.000000   0.000000    0.000000  \n",
       "14143    0.000000   0.000000    0.000000  \n",
       "14144    0.000000   0.000000    0.000000  \n",
       "\n",
       "[14145 rows x 295 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_4 = pd.read_csv(tpm_ghahramani_path, sep=',')\n",
    "print_rowcol('Loaded Ghahramani', df_tpm_4)\n",
    "df_tpm_4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename gene name column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AG1neg_1</th>\n",
       "      <th>AG2pos_2</th>\n",
       "      <th>AG1neg_3</th>\n",
       "      <th>AG2pos_4</th>\n",
       "      <th>AG1neg_5</th>\n",
       "      <th>AG2pos_6</th>\n",
       "      <th>AG1neg_7</th>\n",
       "      <th>AG2pos_8</th>\n",
       "      <th>AG1neg_9</th>\n",
       "      <th>...</th>\n",
       "      <th>GHA501A86</th>\n",
       "      <th>GHA501A88</th>\n",
       "      <th>GHA501A89</th>\n",
       "      <th>GHA501A90</th>\n",
       "      <th>GHA501A93</th>\n",
       "      <th>GHA501A94</th>\n",
       "      <th>GHA501A95</th>\n",
       "      <th>GHA501A98</th>\n",
       "      <th>GHA501A99</th>\n",
       "      <th>GHA501A9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>122.014549</td>\n",
       "      <td>7.484987</td>\n",
       "      <td>597.717004</td>\n",
       "      <td>643.657745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>331.238736</td>\n",
       "      <td>117.133604</td>\n",
       "      <td>468.887523</td>\n",
       "      <td>243.073894</td>\n",
       "      <td>...</td>\n",
       "      <td>61.411894</td>\n",
       "      <td>230.333860</td>\n",
       "      <td>98.040207</td>\n",
       "      <td>93.716396</td>\n",
       "      <td>673.177033</td>\n",
       "      <td>50.220174</td>\n",
       "      <td>303.137672</td>\n",
       "      <td>85.992001</td>\n",
       "      <td>14.915042</td>\n",
       "      <td>4.671418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.560498</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.958870</td>\n",
       "      <td>2.201127</td>\n",
       "      <td>1.425608</td>\n",
       "      <td>222.350975</td>\n",
       "      <td>10.621630</td>\n",
       "      <td>433.142296</td>\n",
       "      <td>1.590745</td>\n",
       "      <td>540.554230</td>\n",
       "      <td>2.841134</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Apoh</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.569181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>174.856508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Narf</td>\n",
       "      <td>1.475207</td>\n",
       "      <td>7.818909</td>\n",
       "      <td>111.868532</td>\n",
       "      <td>117.207838</td>\n",
       "      <td>8.548119</td>\n",
       "      <td>34.838241</td>\n",
       "      <td>49.003077</td>\n",
       "      <td>50.385245</td>\n",
       "      <td>107.228871</td>\n",
       "      <td>...</td>\n",
       "      <td>7.230203</td>\n",
       "      <td>2.514732</td>\n",
       "      <td>1.954466</td>\n",
       "      <td>2.185207</td>\n",
       "      <td>8.604779</td>\n",
       "      <td>3.643098</td>\n",
       "      <td>219.721960</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.016629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cav2</td>\n",
       "      <td>160.239035</td>\n",
       "      <td>278.810852</td>\n",
       "      <td>112.274304</td>\n",
       "      <td>41.820736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28.127109</td>\n",
       "      <td>108.345577</td>\n",
       "      <td>65.333237</td>\n",
       "      <td>4.705999</td>\n",
       "      <td>...</td>\n",
       "      <td>2.163511</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40.207721</td>\n",
       "      <td>1.198789</td>\n",
       "      <td>241.109633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.598203</td>\n",
       "      <td>221.756227</td>\n",
       "      <td>59.831210</td>\n",
       "      <td>7.808018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 295 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name    AG1neg_1    AG2pos_2    AG1neg_3    AG2pos_4  AG1neg_5  \\\n",
       "0     Gnai3  122.014549    7.484987  597.717004  643.657745  0.000000   \n",
       "1     Cdc45    0.000000    0.000000    0.000000    5.560498  0.000000   \n",
       "2      Apoh    0.000000    0.000000    0.000000    0.000000  0.000000   \n",
       "3      Narf    1.475207    7.818909  111.868532  117.207838  8.548119   \n",
       "4      Cav2  160.239035  278.810852  112.274304   41.820736  0.000000   \n",
       "\n",
       "     AG2pos_6    AG1neg_7    AG2pos_8    AG1neg_9  ...  GHA501A86   GHA501A88  \\\n",
       "0  331.238736  117.133604  468.887523  243.073894  ...  61.411894  230.333860   \n",
       "1    0.000000    0.000000    0.000000    0.000000  ...   0.958870    2.201127   \n",
       "2    0.000000    0.000000    0.000000    0.000000  ...   0.000000    0.000000   \n",
       "3   34.838241   49.003077   50.385245  107.228871  ...   7.230203    2.514732   \n",
       "4   28.127109  108.345577   65.333237    4.705999  ...   2.163511    0.000000   \n",
       "\n",
       "   GHA501A89   GHA501A90   GHA501A93   GHA501A94   GHA501A95   GHA501A98  \\\n",
       "0  98.040207   93.716396  673.177033   50.220174  303.137672   85.992001   \n",
       "1   1.425608  222.350975   10.621630  433.142296    1.590745  540.554230   \n",
       "2   0.000000    0.000000    0.000000    0.000000    2.569181    0.000000   \n",
       "3   1.954466    2.185207    8.604779    3.643098  219.721960    0.000000   \n",
       "4  40.207721    1.198789  241.109633    0.000000    0.598203  221.756227   \n",
       "\n",
       "   GHA501A99    GHA501A9  \n",
       "0  14.915042    4.671418  \n",
       "1   2.841134    0.000000  \n",
       "2   0.000000  174.856508  \n",
       "3   0.000000    1.016629  \n",
       "4  59.831210    7.808018  \n",
       "\n",
       "[5 rows x 295 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_4.rename( columns={'Unnamed: 0':'gene_name'}, inplace=True)\n",
    "df_tpm_4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set gene to index after checking unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_4.gene_name.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AG1neg_1</th>\n",
       "      <th>AG2pos_2</th>\n",
       "      <th>AG1neg_3</th>\n",
       "      <th>AG2pos_4</th>\n",
       "      <th>AG1neg_5</th>\n",
       "      <th>AG2pos_6</th>\n",
       "      <th>AG1neg_7</th>\n",
       "      <th>AG2pos_8</th>\n",
       "      <th>AG1neg_9</th>\n",
       "      <th>...</th>\n",
       "      <th>GHA501A86</th>\n",
       "      <th>GHA501A88</th>\n",
       "      <th>GHA501A89</th>\n",
       "      <th>GHA501A90</th>\n",
       "      <th>GHA501A93</th>\n",
       "      <th>GHA501A94</th>\n",
       "      <th>GHA501A95</th>\n",
       "      <th>GHA501A98</th>\n",
       "      <th>GHA501A99</th>\n",
       "      <th>GHA501A9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gnai3</th>\n",
       "      <td>Gnai3</td>\n",
       "      <td>122.014549</td>\n",
       "      <td>7.484987</td>\n",
       "      <td>597.717004</td>\n",
       "      <td>643.657745</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>331.238736</td>\n",
       "      <td>117.133604</td>\n",
       "      <td>468.887523</td>\n",
       "      <td>243.073894</td>\n",
       "      <td>...</td>\n",
       "      <td>61.411894</td>\n",
       "      <td>230.333860</td>\n",
       "      <td>98.040207</td>\n",
       "      <td>93.716396</td>\n",
       "      <td>673.177033</td>\n",
       "      <td>50.220174</td>\n",
       "      <td>303.137672</td>\n",
       "      <td>85.992001</td>\n",
       "      <td>14.915042</td>\n",
       "      <td>4.671418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cdc45</th>\n",
       "      <td>Cdc45</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.560498</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.958870</td>\n",
       "      <td>2.201127</td>\n",
       "      <td>1.425608</td>\n",
       "      <td>222.350975</td>\n",
       "      <td>10.621630</td>\n",
       "      <td>433.142296</td>\n",
       "      <td>1.590745</td>\n",
       "      <td>540.554230</td>\n",
       "      <td>2.841134</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Apoh</th>\n",
       "      <td>Apoh</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.569181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>174.856508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Narf</th>\n",
       "      <td>Narf</td>\n",
       "      <td>1.475207</td>\n",
       "      <td>7.818909</td>\n",
       "      <td>111.868532</td>\n",
       "      <td>117.207838</td>\n",
       "      <td>8.548119</td>\n",
       "      <td>34.838241</td>\n",
       "      <td>49.003077</td>\n",
       "      <td>50.385245</td>\n",
       "      <td>107.228871</td>\n",
       "      <td>...</td>\n",
       "      <td>7.230203</td>\n",
       "      <td>2.514732</td>\n",
       "      <td>1.954466</td>\n",
       "      <td>2.185207</td>\n",
       "      <td>8.604779</td>\n",
       "      <td>3.643098</td>\n",
       "      <td>219.721960</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.016629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cav2</th>\n",
       "      <td>Cav2</td>\n",
       "      <td>160.239035</td>\n",
       "      <td>278.810852</td>\n",
       "      <td>112.274304</td>\n",
       "      <td>41.820736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>28.127109</td>\n",
       "      <td>108.345577</td>\n",
       "      <td>65.333237</td>\n",
       "      <td>4.705999</td>\n",
       "      <td>...</td>\n",
       "      <td>2.163511</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40.207721</td>\n",
       "      <td>1.198789</td>\n",
       "      <td>241.109633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.598203</td>\n",
       "      <td>221.756227</td>\n",
       "      <td>59.831210</td>\n",
       "      <td>7.808018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 295 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name    AG1neg_1    AG2pos_2    AG1neg_3    AG2pos_4  AG1neg_5  \\\n",
       "gene_name                                                                       \n",
       "Gnai3         Gnai3  122.014549    7.484987  597.717004  643.657745  0.000000   \n",
       "Cdc45         Cdc45    0.000000    0.000000    0.000000    5.560498  0.000000   \n",
       "Apoh           Apoh    0.000000    0.000000    0.000000    0.000000  0.000000   \n",
       "Narf           Narf    1.475207    7.818909  111.868532  117.207838  8.548119   \n",
       "Cav2           Cav2  160.239035  278.810852  112.274304   41.820736  0.000000   \n",
       "\n",
       "             AG2pos_6    AG1neg_7    AG2pos_8    AG1neg_9  ...  GHA501A86  \\\n",
       "gene_name                                                  ...              \n",
       "Gnai3      331.238736  117.133604  468.887523  243.073894  ...  61.411894   \n",
       "Cdc45        0.000000    0.000000    0.000000    0.000000  ...   0.958870   \n",
       "Apoh         0.000000    0.000000    0.000000    0.000000  ...   0.000000   \n",
       "Narf        34.838241   49.003077   50.385245  107.228871  ...   7.230203   \n",
       "Cav2        28.127109  108.345577   65.333237    4.705999  ...   2.163511   \n",
       "\n",
       "            GHA501A88  GHA501A89   GHA501A90   GHA501A93   GHA501A94  \\\n",
       "gene_name                                                              \n",
       "Gnai3      230.333860  98.040207   93.716396  673.177033   50.220174   \n",
       "Cdc45        2.201127   1.425608  222.350975   10.621630  433.142296   \n",
       "Apoh         0.000000   0.000000    0.000000    0.000000    0.000000   \n",
       "Narf         2.514732   1.954466    2.185207    8.604779    3.643098   \n",
       "Cav2         0.000000  40.207721    1.198789  241.109633    0.000000   \n",
       "\n",
       "            GHA501A95   GHA501A98  GHA501A99    GHA501A9  \n",
       "gene_name                                                 \n",
       "Gnai3      303.137672   85.992001  14.915042    4.671418  \n",
       "Cdc45        1.590745  540.554230   2.841134    0.000000  \n",
       "Apoh         2.569181    0.000000   0.000000  174.856508  \n",
       "Narf       219.721960    0.000000   0.000000    1.016629  \n",
       "Cav2         0.598203  221.756227  59.831210    7.808018  \n",
       "\n",
       "[5 rows x 295 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_4.index = df_tpm_4['gene_name']\n",
    "df_tpm_4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Insert gene into data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape after merging gene ids: 14081 rows x 294 cols\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AG1neg_1</th>\n",
       "      <th>AG2pos_2</th>\n",
       "      <th>AG1neg_3</th>\n",
       "      <th>AG2pos_4</th>\n",
       "      <th>AG1neg_5</th>\n",
       "      <th>AG2pos_6</th>\n",
       "      <th>AG1neg_7</th>\n",
       "      <th>AG2pos_8</th>\n",
       "      <th>AG1neg_9</th>\n",
       "      <th>AG2pos_10</th>\n",
       "      <th>...</th>\n",
       "      <th>GHA501A86</th>\n",
       "      <th>GHA501A88</th>\n",
       "      <th>GHA501A89</th>\n",
       "      <th>GHA501A90</th>\n",
       "      <th>GHA501A93</th>\n",
       "      <th>GHA501A94</th>\n",
       "      <th>GHA501A95</th>\n",
       "      <th>GHA501A98</th>\n",
       "      <th>GHA501A99</th>\n",
       "      <th>GHA501A9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000007777</th>\n",
       "      <td>428.888448</td>\n",
       "      <td>127.707832</td>\n",
       "      <td>25.495415</td>\n",
       "      <td>84.045924</td>\n",
       "      <td>27.923632</td>\n",
       "      <td>2.147246</td>\n",
       "      <td>235.262386</td>\n",
       "      <td>154.615303</td>\n",
       "      <td>360.786526</td>\n",
       "      <td>235.518628</td>\n",
       "      <td>...</td>\n",
       "      <td>137.416574</td>\n",
       "      <td>351.590275</td>\n",
       "      <td>100.556455</td>\n",
       "      <td>249.840059</td>\n",
       "      <td>261.627303</td>\n",
       "      <td>309.417973</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>427.928611</td>\n",
       "      <td>25.447842</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000042208</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.373457</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17.226978</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.808099</td>\n",
       "      <td>20.186837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.909022</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.576499</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000058706</th>\n",
       "      <td>3.126150</td>\n",
       "      <td>28.478409</td>\n",
       "      <td>5.513106</td>\n",
       "      <td>7.067672</td>\n",
       "      <td>9.057271</td>\n",
       "      <td>42.485137</td>\n",
       "      <td>24.780884</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.112709</td>\n",
       "      <td>10.485246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.391618</td>\n",
       "      <td>2.329740</td>\n",
       "      <td>0.289421</td>\n",
       "      <td>1.051997</td>\n",
       "      <td>36.670875</td>\n",
       "      <td>4.91037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.538591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108680</th>\n",
       "      <td>3.126150</td>\n",
       "      <td>28.478409</td>\n",
       "      <td>5.513106</td>\n",
       "      <td>7.067672</td>\n",
       "      <td>9.057271</td>\n",
       "      <td>42.485137</td>\n",
       "      <td>24.780884</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.112709</td>\n",
       "      <td>10.485246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.391618</td>\n",
       "      <td>2.329740</td>\n",
       "      <td>0.289421</td>\n",
       "      <td>1.051997</td>\n",
       "      <td>36.670875</td>\n",
       "      <td>4.91037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.538591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000060512</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 294 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      AG1neg_1    AG2pos_2   AG1neg_3   AG2pos_4   AG1neg_5  \\\n",
       "gene_id                                                                       \n",
       "ENSMUSG00000007777  428.888448  127.707832  25.495415  84.045924  27.923632   \n",
       "ENSMUSG00000042208    0.000000    2.373457   0.000000  17.226978   0.000000   \n",
       "ENSMUSG00000058706    3.126150   28.478409   5.513106   7.067672   9.057271   \n",
       "ENSMUSG00000108680    3.126150   28.478409   5.513106   7.067672   9.057271   \n",
       "ENSMUSG00000060512    0.000000    0.000000   0.000000   0.000000   0.000000   \n",
       "\n",
       "                     AG2pos_6    AG1neg_7    AG2pos_8    AG1neg_9   AG2pos_10  \\\n",
       "gene_id                                                                         \n",
       "ENSMUSG00000007777   2.147246  235.262386  154.615303  360.786526  235.518628   \n",
       "ENSMUSG00000042208   0.000000    4.808099   20.186837    0.000000    0.000000   \n",
       "ENSMUSG00000058706  42.485137   24.780884    0.000000    5.112709   10.485246   \n",
       "ENSMUSG00000108680  42.485137   24.780884    0.000000    5.112709   10.485246   \n",
       "ENSMUSG00000060512   0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "                    ...   GHA501A86   GHA501A88   GHA501A89   GHA501A90  \\\n",
       "gene_id             ...                                                   \n",
       "ENSMUSG00000007777  ...  137.416574  351.590275  100.556455  249.840059   \n",
       "ENSMUSG00000042208  ...    0.000000    0.000000    0.000000    0.000000   \n",
       "ENSMUSG00000058706  ...    0.000000   10.391618    2.329740    0.289421   \n",
       "ENSMUSG00000108680  ...    0.000000   10.391618    2.329740    0.289421   \n",
       "ENSMUSG00000060512  ...    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "                     GHA501A93   GHA501A94  GHA501A95   GHA501A98  GHA501A99  \\\n",
       "gene_id                                                                        \n",
       "ENSMUSG00000007777  261.627303  309.417973    0.00000  427.928611  25.447842   \n",
       "ENSMUSG00000042208    0.000000  120.909022    0.00000    0.000000   1.576499   \n",
       "ENSMUSG00000058706    1.051997   36.670875    4.91037    0.000000   0.000000   \n",
       "ENSMUSG00000108680    1.051997   36.670875    4.91037    0.000000   0.000000   \n",
       "ENSMUSG00000060512    0.000000    0.000000    0.00000    0.000000   0.000000   \n",
       "\n",
       "                    GHA501A9  \n",
       "gene_id                       \n",
       "ENSMUSG00000007777  0.000000  \n",
       "ENSMUSG00000042208  0.000000  \n",
       "ENSMUSG00000058706  0.538591  \n",
       "ENSMUSG00000108680  0.538591  \n",
       "ENSMUSG00000060512  0.000000  \n",
       "\n",
       "[5 rows x 294 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_4 = df_tpm_4.drop('gene_name', axis=1)\n",
    "df_tpm_4 = df_gene_name2id.join(df_tpm_4, lsuffix='', rsuffix='', how='inner')\n",
    "df_tpm_4.index = df_tpm_4.gene_id\n",
    "df_tpm_4 = df_tpm_4.drop('gene_id', axis=1)\n",
    "print_rowcol('Shape after merging gene ids', df_tpm_4)\n",
    "df_tpm_4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create merged dataset\n",
    "\n",
    "Create merged dataset from all subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46609, 782)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_combined = df_tpm_1.join(df_tpm_2, lsuffix='', rsuffix='_other', how='inner')\n",
    "df_tpm_combined.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21016, 2204)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_combined = df_tpm_combined.join(df_tpm_3, lsuffix='', rsuffix='_other', how='inner')\n",
    "df_tpm_combined.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13227, 2498)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_combined = df_tpm_combined.join(df_tpm_4, lsuffix='', rsuffix='_other', how='inner')\n",
    "df_tpm_combined.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create `log2(TPM+1)` dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>P1-3-A6</th>\n",
       "      <th>P1-3-A7</th>\n",
       "      <th>...</th>\n",
       "      <th>GHA501A86</th>\n",
       "      <th>GHA501A88</th>\n",
       "      <th>GHA501A89</th>\n",
       "      <th>GHA501A90</th>\n",
       "      <th>GHA501A93</th>\n",
       "      <th>GHA501A94</th>\n",
       "      <th>GHA501A95</th>\n",
       "      <th>GHA501A98</th>\n",
       "      <th>GHA501A99</th>\n",
       "      <th>GHA501A9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.217231</td>\n",
       "      <td>3.003602</td>\n",
       "      <td>4.209453</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.296824</td>\n",
       "      <td>5.300856</td>\n",
       "      <td>5.587965</td>\n",
       "      <td>3.826803</td>\n",
       "      <td>3.414136</td>\n",
       "      <td>...</td>\n",
       "      <td>5.963749</td>\n",
       "      <td>7.853833</td>\n",
       "      <td>6.629942</td>\n",
       "      <td>6.565542</td>\n",
       "      <td>9.396984</td>\n",
       "      <td>5.678640</td>\n",
       "      <td>8.248581</td>\n",
       "      <td>6.442811</td>\n",
       "      <td>3.992319</td>\n",
       "      <td>2.503710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.970022</td>\n",
       "      <td>1.678580</td>\n",
       "      <td>1.278347</td>\n",
       "      <td>7.803169</td>\n",
       "      <td>3.538741</td>\n",
       "      <td>8.762024</td>\n",
       "      <td>1.373367</td>\n",
       "      <td>9.080962</td>\n",
       "      <td>1.941532</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.835593</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.458255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.109361</td>\n",
       "      <td>5.938286</td>\n",
       "      <td>5.093391</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.693766</td>\n",
       "      <td>4.627607</td>\n",
       "      <td>6.537141</td>\n",
       "      <td>3.842979</td>\n",
       "      <td>...</td>\n",
       "      <td>3.040928</td>\n",
       "      <td>1.813415</td>\n",
       "      <td>1.562897</td>\n",
       "      <td>1.671387</td>\n",
       "      <td>3.263752</td>\n",
       "      <td>2.215088</td>\n",
       "      <td>7.786086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.011946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.778209</td>\n",
       "      <td>2.601697</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.143842</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.661527</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.364843</td>\n",
       "      <td>1.136709</td>\n",
       "      <td>7.919517</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.676451</td>\n",
       "      <td>7.799322</td>\n",
       "      <td>5.926740</td>\n",
       "      <td>3.138817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000106918</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.926711</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.521679</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.384395</td>\n",
       "      <td>...</td>\n",
       "      <td>3.217381</td>\n",
       "      <td>4.115454</td>\n",
       "      <td>1.121270</td>\n",
       "      <td>3.643718</td>\n",
       "      <td>4.947408</td>\n",
       "      <td>6.036318</td>\n",
       "      <td>1.982202</td>\n",
       "      <td>4.864736</td>\n",
       "      <td>5.896306</td>\n",
       "      <td>4.078610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000107283</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.424922</td>\n",
       "      <td>2.443607</td>\n",
       "      <td>1.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.503349</td>\n",
       "      <td>3.740928</td>\n",
       "      <td>5.351558</td>\n",
       "      <td>3.264536</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.267812</td>\n",
       "      <td>4.150363</td>\n",
       "      <td>4.719657</td>\n",
       "      <td>6.472461</td>\n",
       "      <td>5.712798</td>\n",
       "      <td>1.906817</td>\n",
       "      <td>5.473054</td>\n",
       "      <td>6.433896</td>\n",
       "      <td>5.444845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000107417</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000107499</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000107645</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.104478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13227 rows × 2498 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    P1-3-A1  P1-3-A10  P1-3-A11  P1-3-A12   P1-3-A2   P1-3-A3  \\\n",
       "ENSMUSG00000000001      0.0  4.217231  3.003602  4.209453  0.000000  5.296824   \n",
       "ENSMUSG00000000028      0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000000049      0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000000056      0.0  3.109361  5.938286  5.093391  0.000000  0.000000   \n",
       "ENSMUSG00000000058      0.0  1.778209  2.601697  0.000000  7.143842  0.000000   \n",
       "...                     ...       ...       ...       ...       ...       ...   \n",
       "ENSMUSG00000106918      0.0  5.926711  0.000000  4.521679  0.000000  0.000000   \n",
       "ENSMUSG00000107283      0.0  4.424922  2.443607  1.584963  0.000000  0.000000   \n",
       "ENSMUSG00000107417      0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000107499      0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000107645      0.0  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                     P1-3-A4   P1-3-A5   P1-3-A6   P1-3-A7  ...  GHA501A86  \\\n",
       "ENSMUSG00000000001  5.300856  5.587965  3.826803  3.414136  ...   5.963749   \n",
       "ENSMUSG00000000028  0.000000  0.000000  0.000000  0.000000  ...   0.970022   \n",
       "ENSMUSG00000000049  0.000000  0.000000  0.000000  0.000000  ...   0.000000   \n",
       "ENSMUSG00000000056  2.693766  4.627607  6.537141  3.842979  ...   3.040928   \n",
       "ENSMUSG00000000058  0.000000  0.000000  0.000000  0.000000  ...   1.661527   \n",
       "...                      ...       ...       ...       ...  ...        ...   \n",
       "ENSMUSG00000106918  4.028569  0.000000  0.000000  5.384395  ...   3.217381   \n",
       "ENSMUSG00000107283  4.503349  3.740928  5.351558  3.264536  ...   0.000000   \n",
       "ENSMUSG00000107417  0.000000  0.000000  0.000000  0.000000  ...   0.000000   \n",
       "ENSMUSG00000107499  0.000000  0.000000  0.000000  0.000000  ...   0.000000   \n",
       "ENSMUSG00000107645  0.000000  0.000000  0.000000  0.000000  ...   0.000000   \n",
       "\n",
       "                    GHA501A88  GHA501A89  GHA501A90  GHA501A93  GHA501A94  \\\n",
       "ENSMUSG00000000001   7.853833   6.629942   6.565542   9.396984   5.678640   \n",
       "ENSMUSG00000000028   1.678580   1.278347   7.803169   3.538741   8.762024   \n",
       "ENSMUSG00000000049   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "ENSMUSG00000000056   1.813415   1.562897   1.671387   3.263752   2.215088   \n",
       "ENSMUSG00000000058   0.000000   5.364843   1.136709   7.919517   0.000000   \n",
       "...                       ...        ...        ...        ...        ...   \n",
       "ENSMUSG00000106918   4.115454   1.121270   3.643718   4.947408   6.036318   \n",
       "ENSMUSG00000107283   4.267812   4.150363   4.719657   6.472461   5.712798   \n",
       "ENSMUSG00000107417   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "ENSMUSG00000107499   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "ENSMUSG00000107645   0.000000   0.000000   0.000000   0.000000   4.104478   \n",
       "\n",
       "                    GHA501A95  GHA501A98  GHA501A99  GHA501A9  \n",
       "ENSMUSG00000000001   8.248581   6.442811   3.992319  2.503710  \n",
       "ENSMUSG00000000028   1.373367   9.080962   1.941532  0.000000  \n",
       "ENSMUSG00000000049   1.835593   0.000000   0.000000  7.458255  \n",
       "ENSMUSG00000000056   7.786086   0.000000   0.000000  1.011946  \n",
       "ENSMUSG00000000058   0.676451   7.799322   5.926740  3.138817  \n",
       "...                       ...        ...        ...       ...  \n",
       "ENSMUSG00000106918   1.982202   4.864736   5.896306  4.078610  \n",
       "ENSMUSG00000107283   1.906817   5.473054   6.433896  5.444845  \n",
       "ENSMUSG00000107417   0.000000   0.000000   0.000000  0.000000  \n",
       "ENSMUSG00000107499   0.000000   0.000000   0.000000  0.000000  \n",
       "ENSMUSG00000107645   0.000000   0.000000   0.000000  0.000000  \n",
       "\n",
       "[13227 rows x 2498 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ltpm_combined = np.log2(df_tpm_combined + 1)\n",
    "df_ltpm_combined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter for genes and cells which match filtering criteria. Call `.shape` to check what was filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6807, 2498)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expression_filt_mask = df_ltpm_combined > min_ltpm_exp\n",
    "df_ltpm_combined_genefilt = df_ltpm_combined[df_expression_filt_mask.sum(axis=1) > min_num_cells_for_gene_exp]\n",
    "df_ltpm_combined_genefilt.shape\n",
    "\n",
    "#df_expression_filt_mask = df_ltpm_combined > 0\n",
    "#df_ltpm_combined_genefilt = df_ltpm_combined[df_expression_filt_mask.sum(axis=1) > 1000]\n",
    "#df_ltpm_combined_genefilt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6807, 2298)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ltpm_combined_genecellfilt = df_ltpm_combined_genefilt \\\n",
    "    .T[(df_ltpm_combined_genefilt > min_ltpm_exp).sum(axis=0) > min_num_genes_in_cell_exp]\n",
    "df_ltpm_combined_genecellfilt = df_ltpm_combined_genecellfilt.T\n",
    "df_ltpm_combined_genecellfilt.shape\n",
    "\n",
    "#df_ltpm_combined_genecellfilt = df_ltpm_combined_genefilt \\\n",
    "#    .T[(df_ltpm_combined_genefilt > 1).sum(axis=0) > 500]\n",
    "#df_ltpm_combined_genecellfilt = df_ltpm_combined_genecellfilt.T\n",
    "#df_ltpm_combined_genecellfilt.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalise data\n",
    "\n",
    "Check max values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18.320341299017063\n"
     ]
    }
   ],
   "source": [
    "data_max = df_ltpm_combined_genecellfilt.max()\n",
    "data_max = data_max.max()\n",
    "print(data_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Normalise data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MinMaxScaler(copy=True, feature_range=(0, 1))\n",
      "(6807,)\n"
     ]
    }
   ],
   "source": [
    "np_data = df_ltpm_combined_genecellfilt.T.values\n",
    "scaler = MinMaxScaler()\n",
    "print(scaler.fit(np_data))\n",
    "\n",
    "# Check which dimension we are fitting to - if we are fitting to gene expression then should be equal to number of genes\n",
    "print(scaler.data_max_.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6807, 2298)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np_data_norm = np.transpose(scaler.transform(np_data))\n",
    "np_data_norm.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>P1-3-A6</th>\n",
       "      <th>P1-3-A7</th>\n",
       "      <th>P1-3-A8</th>\n",
       "      <th>...</th>\n",
       "      <th>GHA501A86</th>\n",
       "      <th>GHA501A88</th>\n",
       "      <th>GHA501A89</th>\n",
       "      <th>GHA501A90</th>\n",
       "      <th>GHA501A93</th>\n",
       "      <th>GHA501A94</th>\n",
       "      <th>GHA501A95</th>\n",
       "      <th>GHA501A98</th>\n",
       "      <th>GHA501A99</th>\n",
       "      <th>GHA501A9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.405649</td>\n",
       "      <td>0.288912</td>\n",
       "      <td>0.404901</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.509494</td>\n",
       "      <td>0.509881</td>\n",
       "      <td>0.537498</td>\n",
       "      <td>0.368094</td>\n",
       "      <td>0.328401</td>\n",
       "      <td>0.470171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.573644</td>\n",
       "      <td>0.755448</td>\n",
       "      <td>0.637724</td>\n",
       "      <td>0.631530</td>\n",
       "      <td>0.903882</td>\n",
       "      <td>0.546220</td>\n",
       "      <td>0.793419</td>\n",
       "      <td>0.619724</td>\n",
       "      <td>0.384015</td>\n",
       "      <td>0.240828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000056</th>\n",
       "      <td>0.354969</td>\n",
       "      <td>0.677923</td>\n",
       "      <td>0.581469</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.307524</td>\n",
       "      <td>0.528294</td>\n",
       "      <td>0.746289</td>\n",
       "      <td>0.438720</td>\n",
       "      <td>0.318122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.347157</td>\n",
       "      <td>0.207022</td>\n",
       "      <td>0.178423</td>\n",
       "      <td>0.190808</td>\n",
       "      <td>0.372595</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.888871</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <td>0.184017</td>\n",
       "      <td>0.269236</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.739278</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.171943</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.555179</td>\n",
       "      <td>0.117632</td>\n",
       "      <td>0.819548</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.070002</td>\n",
       "      <td>0.807110</td>\n",
       "      <td>0.613327</td>\n",
       "      <td>0.324819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000078</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.265407</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.049705</td>\n",
       "      <td>0.280146</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.130612</td>\n",
       "      <td>0.076877</td>\n",
       "      <td>...</td>\n",
       "      <td>0.569237</td>\n",
       "      <td>0.538678</td>\n",
       "      <td>0.584073</td>\n",
       "      <td>0.598017</td>\n",
       "      <td>0.254595</td>\n",
       "      <td>0.694776</td>\n",
       "      <td>0.500197</td>\n",
       "      <td>0.189335</td>\n",
       "      <td>0.787864</td>\n",
       "      <td>0.565605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000085</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.557392</td>\n",
       "      <td>0.442670</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.513107</td>\n",
       "      <td>0.662761</td>\n",
       "      <td>0.228048</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083766</td>\n",
       "      <td>0.501817</td>\n",
       "      <td>0.286331</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.367558</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000101892</th>\n",
       "      <td>0.500886</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.315071</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.427417</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.498924</td>\n",
       "      <td>0.148768</td>\n",
       "      <td>0.549690</td>\n",
       "      <td>0.506464</td>\n",
       "      <td>0.668826</td>\n",
       "      <td>0.623066</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.817074</td>\n",
       "      <td>0.344403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102976</th>\n",
       "      <td>0.426120</td>\n",
       "      <td>0.260469</td>\n",
       "      <td>0.425397</td>\n",
       "      <td>0.490006</td>\n",
       "      <td>0.504198</td>\n",
       "      <td>0.362395</td>\n",
       "      <td>0.373281</td>\n",
       "      <td>0.483482</td>\n",
       "      <td>0.383271</td>\n",
       "      <td>0.485597</td>\n",
       "      <td>...</td>\n",
       "      <td>0.490887</td>\n",
       "      <td>0.421085</td>\n",
       "      <td>0.400355</td>\n",
       "      <td>0.352684</td>\n",
       "      <td>0.356345</td>\n",
       "      <td>0.624218</td>\n",
       "      <td>0.459310</td>\n",
       "      <td>0.362824</td>\n",
       "      <td>0.124659</td>\n",
       "      <td>0.537819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000106864</th>\n",
       "      <td>0.457133</td>\n",
       "      <td>0.329494</td>\n",
       "      <td>0.197375</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.656869</td>\n",
       "      <td>0.320657</td>\n",
       "      <td>0.538464</td>\n",
       "      <td>0.800126</td>\n",
       "      <td>0.241875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.378541</td>\n",
       "      <td>0.027108</td>\n",
       "      <td>0.325596</td>\n",
       "      <td>0.168844</td>\n",
       "      <td>0.110277</td>\n",
       "      <td>0.461184</td>\n",
       "      <td>0.112672</td>\n",
       "      <td>0.424511</td>\n",
       "      <td>0.222422</td>\n",
       "      <td>0.281608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000106918</th>\n",
       "      <td>0.736540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.561930</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500649</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.669144</td>\n",
       "      <td>0.599597</td>\n",
       "      <td>...</td>\n",
       "      <td>0.399839</td>\n",
       "      <td>0.511447</td>\n",
       "      <td>0.139346</td>\n",
       "      <td>0.452822</td>\n",
       "      <td>0.614838</td>\n",
       "      <td>0.750162</td>\n",
       "      <td>0.246338</td>\n",
       "      <td>0.604564</td>\n",
       "      <td>0.732762</td>\n",
       "      <td>0.506868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000107283</th>\n",
       "      <td>0.538283</td>\n",
       "      <td>0.297260</td>\n",
       "      <td>0.192807</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.547823</td>\n",
       "      <td>0.455076</td>\n",
       "      <td>0.651006</td>\n",
       "      <td>0.397124</td>\n",
       "      <td>0.603628</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.519171</td>\n",
       "      <td>0.504883</td>\n",
       "      <td>0.574137</td>\n",
       "      <td>0.787362</td>\n",
       "      <td>0.694950</td>\n",
       "      <td>0.231960</td>\n",
       "      <td>0.665786</td>\n",
       "      <td>0.782670</td>\n",
       "      <td>0.662354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6807 rows × 2298 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    P1-3-A10  P1-3-A11  P1-3-A12   P1-3-A2   P1-3-A3  \\\n",
       "ENSMUSG00000000001  0.405649  0.288912  0.404901  0.000000  0.509494   \n",
       "ENSMUSG00000000056  0.354969  0.677923  0.581469  0.000000  0.000000   \n",
       "ENSMUSG00000000058  0.184017  0.269236  0.000000  0.739278  0.000000   \n",
       "ENSMUSG00000000078  0.000000  0.000000  0.265407  0.000000  0.000000   \n",
       "ENSMUSG00000000085  0.000000  0.557392  0.442670  0.000000  0.000000   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "ENSMUSG00000101892  0.500886  0.000000  0.315071  0.000000  0.000000   \n",
       "ENSMUSG00000102976  0.426120  0.260469  0.425397  0.490006  0.504198   \n",
       "ENSMUSG00000106864  0.457133  0.329494  0.197375  0.000000  0.000000   \n",
       "ENSMUSG00000106918  0.736540  0.000000  0.561930  0.000000  0.000000   \n",
       "ENSMUSG00000107283  0.538283  0.297260  0.192807  0.000000  0.000000   \n",
       "\n",
       "                     P1-3-A4   P1-3-A5   P1-3-A6   P1-3-A7   P1-3-A8  ...  \\\n",
       "ENSMUSG00000000001  0.509881  0.537498  0.368094  0.328401  0.470171  ...   \n",
       "ENSMUSG00000000056  0.307524  0.528294  0.746289  0.438720  0.318122  ...   \n",
       "ENSMUSG00000000058  0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSMUSG00000000078  0.049705  0.280146  0.000000  0.130612  0.076877  ...   \n",
       "ENSMUSG00000000085  0.000000  0.513107  0.662761  0.228048  0.000000  ...   \n",
       "...                      ...       ...       ...       ...       ...  ...   \n",
       "ENSMUSG00000101892  0.427417  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSMUSG00000102976  0.362395  0.373281  0.483482  0.383271  0.485597  ...   \n",
       "ENSMUSG00000106864  0.656869  0.320657  0.538464  0.800126  0.241875  ...   \n",
       "ENSMUSG00000106918  0.500649  0.000000  0.000000  0.669144  0.599597  ...   \n",
       "ENSMUSG00000107283  0.547823  0.455076  0.651006  0.397124  0.603628  ...   \n",
       "\n",
       "                    GHA501A86  GHA501A88  GHA501A89  GHA501A90  GHA501A93  \\\n",
       "ENSMUSG00000000001   0.573644   0.755448   0.637724   0.631530   0.903882   \n",
       "ENSMUSG00000000056   0.347157   0.207022   0.178423   0.190808   0.372595   \n",
       "ENSMUSG00000000058   0.171943   0.000000   0.555179   0.117632   0.819548   \n",
       "ENSMUSG00000000078   0.569237   0.538678   0.584073   0.598017   0.254595   \n",
       "ENSMUSG00000000085   0.083766   0.501817   0.286331   0.000000   0.000000   \n",
       "...                       ...        ...        ...        ...        ...   \n",
       "ENSMUSG00000101892   0.000000   0.498924   0.148768   0.549690   0.506464   \n",
       "ENSMUSG00000102976   0.490887   0.421085   0.400355   0.352684   0.356345   \n",
       "ENSMUSG00000106864   0.378541   0.027108   0.325596   0.168844   0.110277   \n",
       "ENSMUSG00000106918   0.399839   0.511447   0.139346   0.452822   0.614838   \n",
       "ENSMUSG00000107283   0.000000   0.519171   0.504883   0.574137   0.787362   \n",
       "\n",
       "                    GHA501A94  GHA501A95  GHA501A98  GHA501A99  GHA501A9  \n",
       "ENSMUSG00000000001   0.546220   0.793419   0.619724   0.384015  0.240828  \n",
       "ENSMUSG00000000056   0.252878   0.888871   0.000000   0.000000  0.115525  \n",
       "ENSMUSG00000000058   0.000000   0.070002   0.807110   0.613327  0.324819  \n",
       "ENSMUSG00000000078   0.694776   0.500197   0.189335   0.787864  0.565605  \n",
       "ENSMUSG00000000085   0.000000   0.000000   0.000000   0.367558  0.000000  \n",
       "...                       ...        ...        ...        ...       ...  \n",
       "ENSMUSG00000101892   0.668826   0.623066   0.000000   0.817074  0.344403  \n",
       "ENSMUSG00000102976   0.624218   0.459310   0.362824   0.124659  0.537819  \n",
       "ENSMUSG00000106864   0.461184   0.112672   0.424511   0.222422  0.281608  \n",
       "ENSMUSG00000106918   0.750162   0.246338   0.604564   0.732762  0.506868  \n",
       "ENSMUSG00000107283   0.694950   0.231960   0.665786   0.782670  0.662354  \n",
       "\n",
       "[6807 rows x 2298 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ltpm_combined_norm = pd.DataFrame(np_data_norm)\n",
    "df_ltpm_combined_norm.columns = df_ltpm_combined_genecellfilt.columns\n",
    "df_ltpm_combined_norm.index = df_ltpm_combined_genecellfilt.index\n",
    "df_ltpm_combined_norm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check new max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0\n"
     ]
    }
   ],
   "source": [
    "data_max = df_ltpm_combined_norm.max()\n",
    "data_max = data_max.max()\n",
    "print(data_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Split train and test data sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created training dataset: 6807 rows x 1798 cols\n",
      "Created test dataset: 6807 rows x 500 cols\n"
     ]
    }
   ],
   "source": [
    "# Randomly select test and training data\n",
    "train_features, test_features = train_test_split(df_ltpm_combined_norm.T, test_size=test_data_size)\n",
    "train_features = train_features.T\n",
    "test_features = test_features.T\n",
    "\n",
    "print_rowcol('Created training dataset', train_features)\n",
    "print_rowcol('Created test dataset', test_features)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write data to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the column and row names as a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1798, 1)\n",
      "(6807, 1)\n",
      "(500, 1)\n",
      "(6807, 1)\n"
     ]
    }
   ],
   "source": [
    "train_df_column_names = pd.DataFrame(list(train_features.columns.values))\n",
    "train_df_row_names = pd.DataFrame(list(train_features.index.values))\n",
    "test_df_column_names = pd.DataFrame(list(test_features.columns.values))\n",
    "test_df_row_names = pd.DataFrame(list(test_features.index.values))\n",
    "\n",
    "print(train_df_column_names.shape)\n",
    "print(train_df_row_names.shape)\n",
    "print(test_df_column_names.shape)\n",
    "print(test_df_row_names.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the data to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_features.to_csv(data_path + '/tpm_combined.csv', index=False, header=False)\n",
    "train_df_column_names.to_csv(data_path + '/tpm_combined_cols.csv', index=False, header=False)\n",
    "train_df_row_names.to_csv(data_path + '/tpm_combined_rows.csv', index=False, header=False)\n",
    "\n",
    "test_features.to_csv(data_path + '/tpm_combined_test.csv', index=False, header=False)\n",
    "test_df_column_names.to_csv(data_path + '/tpm_combined_cols_test.csv', index=False, header=False)\n",
    "test_df_row_names.to_csv(data_path + '/tpm_combined_rows_test.csv', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
