{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup -CHECK SPELLING MISTAKES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install a pip package in the current Jupyter kernel - REMOVE IN PROD\n",
    "#import sys\n",
    "#!{sys.executable} -m pip install gtfparse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import relevant libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libs\n",
    "import os.path\n",
    "from os import path\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import urllib.request\n",
    "from gtfparse import read_gtf\n",
    "from biomart import BiomartServer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setup paths - main data path is folder inside the code repository where the data dowloaded from download-data has been stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data'\n",
    "gtf_path_gz = data_path + '/Mus_musculus.GRCm38.99.gtf.gz'\n",
    "gtf_path = data_path + '/Mus_musculus.GRCm38.99.gtf'\n",
    "tpm_yang_path_gz = data_path + '/GSE90848_Ana6_basal_hair_bulb_TPM.txt.gz'\n",
    "tpm_yang_path = data_path + '/GSE90848_Ana6_basal_hair_bulb_TPM.txt'\n",
    "gene_tsv_path = data_path + '/gene_names.tsv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set the main thresholds for the pre-processing\n",
    "- min_num_genes_in_cell_exp - is the min number of genes that need to have expression > **min_ltpm_exp** for it to be considered a valid cell\n",
    "- min_num_cells_for_gene_exp - is the min number of cells that need to have an expression for that gene > **min_ltpm_exp** for it to be considered a valid gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thresholds\n",
    "min_num_genes_in_cell_exp = 1000\n",
    "min_num_cells_for_gene_exp = 10\n",
    "min_ltpm_exp = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Resolve Gene names\n",
    "\n",
    "Load GTF data and show some data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>havana</td>\n",
       "      <td>TEC</td>\n",
       "      <td>basic</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001343744</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NA</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868325</th>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>708</td>\n",
       "      <td>752</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868326</th>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSMUSE00000997159</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868327</th>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>565</td>\n",
       "      <td>633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868328</th>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSMUSE00001007635</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868329</th>\n",
       "      <td>JH584295.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>CDS</td>\n",
       "      <td>66</td>\n",
       "      <td>109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095742</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSMUSP00000137004</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1868330 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname   source     feature    start      end  score strand  \\\n",
       "0                 1   havana        gene  3073253  3074322    NaN      +   \n",
       "1                 1   havana  transcript  3073253  3074322    NaN      +   \n",
       "2                 1   havana        exon  3073253  3074322    NaN      +   \n",
       "3                 1  ensembl        gene  3102016  3102125    NaN      +   \n",
       "4                 1  ensembl  transcript  3102016  3102125    NaN      +   \n",
       "...             ...      ...         ...      ...      ...    ...    ...   \n",
       "1868325  JH584295.1  ensembl         CDS      708      752    NaN      -   \n",
       "1868326  JH584295.1  ensembl        exon      565      633    NaN      -   \n",
       "1868327  JH584295.1  ensembl         CDS      565      633    NaN      -   \n",
       "1868328  JH584295.1  ensembl        exon       66      109    NaN      -   \n",
       "1868329  JH584295.1  ensembl         CDS       66      109    NaN      -   \n",
       "\n",
       "        frame             gene_id gene_version  ... transcript_source  \\\n",
       "0           0  ENSMUSG00000102693            1  ...                     \n",
       "1           0  ENSMUSG00000102693            1  ...            havana   \n",
       "2           0  ENSMUSG00000102693            1  ...            havana   \n",
       "3           0  ENSMUSG00000064842            1  ...                     \n",
       "4           0  ENSMUSG00000064842            1  ...           ensembl   \n",
       "...       ...                 ...          ...  ...               ...   \n",
       "1868325     2  ENSMUSG00000095742            1  ...           ensembl   \n",
       "1868326     0  ENSMUSG00000095742            1  ...           ensembl   \n",
       "1868327     2  ENSMUSG00000095742            1  ...           ensembl   \n",
       "1868328     0  ENSMUSG00000095742            1  ...           ensembl   \n",
       "1868329     2  ENSMUSG00000095742            1  ...           ensembl   \n",
       "\n",
       "        transcript_biotype    tag transcript_support_level exon_number  \\\n",
       "0                                                                        \n",
       "1                      TEC  basic                       NA               \n",
       "2                      TEC  basic                       NA           1   \n",
       "3                                                                        \n",
       "4                    snRNA  basic                       NA               \n",
       "...                    ...    ...                      ...         ...   \n",
       "1868325     protein_coding  basic                        5           5   \n",
       "1868326     protein_coding  basic                        5           6   \n",
       "1868327     protein_coding  basic                        5           6   \n",
       "1868328     protein_coding  basic                        5           7   \n",
       "1868329     protein_coding  basic                        5           7   \n",
       "\n",
       "                    exon_id exon_version ccds_id          protein_id  \\\n",
       "0                                                                      \n",
       "1                                                                      \n",
       "2        ENSMUSE00001343744            1                               \n",
       "3                                                                      \n",
       "4                                                                      \n",
       "...                     ...          ...     ...                 ...   \n",
       "1868325                                           ENSMUSP00000137004   \n",
       "1868326  ENSMUSE00000997159            1                               \n",
       "1868327                                           ENSMUSP00000137004   \n",
       "1868328  ENSMUSE00001007635            1                               \n",
       "1868329                                           ENSMUSP00000137004   \n",
       "\n",
       "        protein_version  \n",
       "0                        \n",
       "1                        \n",
       "2                        \n",
       "3                        \n",
       "4                        \n",
       "...                 ...  \n",
       "1868325               1  \n",
       "1868326                  \n",
       "1868327               1  \n",
       "1868328                  \n",
       "1868329               1  \n",
       "\n",
       "[1868330 rows x 26 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf = read_gtf(gtf_path)\n",
    "df_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter GTF data for exons, 3' and 5' UTRs, call `.shape` to see how many rows we have"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1025718, 26)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gtf_filt = df_gtf[(df_gtf.feature=='exon') | (df_gtf.feature=='three_prime_utr') | (df_gtf.feature=='five_prime_utr')]\n",
    "df_gtf_filt.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load in RNA-Seq data\n",
    "\n",
    "Read the yang1 data set and look at the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>P1-3-A6</th>\n",
       "      <th>...</th>\n",
       "      <th>P2-1-H11</th>\n",
       "      <th>P2-1-H12</th>\n",
       "      <th>P2-1-H2</th>\n",
       "      <th>P2-1-H3</th>\n",
       "      <th>P2-1-H4</th>\n",
       "      <th>P2-1-H5</th>\n",
       "      <th>P2-1-H6</th>\n",
       "      <th>P2-1-H7</th>\n",
       "      <th>P2-1-H8</th>\n",
       "      <th>P2-1-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001_Gnai3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.60</td>\n",
       "      <td>7.02</td>\n",
       "      <td>17.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>38.31</td>\n",
       "      <td>38.42</td>\n",
       "      <td>47.10</td>\n",
       "      <td>13.19</td>\n",
       "      <td>...</td>\n",
       "      <td>45.27</td>\n",
       "      <td>5.70</td>\n",
       "      <td>13.80</td>\n",
       "      <td>15.89</td>\n",
       "      <td>19.42</td>\n",
       "      <td>3.21</td>\n",
       "      <td>5.65</td>\n",
       "      <td>20.33</td>\n",
       "      <td>16.72</td>\n",
       "      <td>33.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003_Pbsn</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028_Cdc45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.55</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031_H19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>112.13</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037_Scml2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>ERCC-00164</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46605</th>\n",
       "      <td>ERCC-00165</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>42.44</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.45</td>\n",
       "      <td>9.12</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>5.41</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>13.15</td>\n",
       "      <td>40.04</td>\n",
       "      <td>9.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>18.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46606</th>\n",
       "      <td>ERCC-00168</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46607</th>\n",
       "      <td>ERCC-00170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46608</th>\n",
       "      <td>ERCC-00171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>143.91</td>\n",
       "      <td>1398.07</td>\n",
       "      <td>328.98</td>\n",
       "      <td>417.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>221.50</td>\n",
       "      <td>695.75</td>\n",
       "      <td>632.11</td>\n",
       "      <td>...</td>\n",
       "      <td>485.62</td>\n",
       "      <td>469.37</td>\n",
       "      <td>290.69</td>\n",
       "      <td>135.58</td>\n",
       "      <td>196.35</td>\n",
       "      <td>388.55</td>\n",
       "      <td>357.51</td>\n",
       "      <td>83.07</td>\n",
       "      <td>220.80</td>\n",
       "      <td>408.54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 385 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Gene_id  P1-3-A1  P1-3-A10  P1-3-A11  P1-3-A12  \\\n",
       "0      ENSMUSG00000000001_Gnai3      0.0     17.60      7.02     17.50   \n",
       "1       ENSMUSG00000000003_Pbsn      0.0      0.00      0.00      0.00   \n",
       "2      ENSMUSG00000000028_Cdc45      0.0      0.00      0.00      0.00   \n",
       "3        ENSMUSG00000000031_H19      0.0      0.00      0.00      0.00   \n",
       "4      ENSMUSG00000000037_Scml2      0.0      0.00      0.00      0.00   \n",
       "...                         ...      ...       ...       ...       ...   \n",
       "46604                ERCC-00164      0.0      0.00      0.00      0.00   \n",
       "46605                ERCC-00165      0.0      0.00     42.44      3.28   \n",
       "46606                ERCC-00168      0.0      0.00      0.00      0.00   \n",
       "46607                ERCC-00170      0.0      0.00      0.00      0.00   \n",
       "46608                ERCC-00171      0.0    143.91   1398.07    328.98   \n",
       "\n",
       "       P1-3-A2  P1-3-A3  P1-3-A4  P1-3-A5  P1-3-A6  ...  P2-1-H11  P2-1-H12  \\\n",
       "0         0.00    38.31    38.42    47.10    13.19  ...     45.27      5.70   \n",
       "1         0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "2         0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "3         0.00     0.00     0.00     0.00   112.13  ...      0.00      6.80   \n",
       "4         0.00     0.00     0.00     0.00     0.00  ...      1.47      0.00   \n",
       "...        ...      ...      ...      ...      ...  ...       ...       ...   \n",
       "46604     0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "46605     0.00     0.00     9.45     9.12     0.00  ...      5.41      1.76   \n",
       "46606     0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "46607     0.00     0.00     0.00     0.00     0.00  ...      0.00      0.00   \n",
       "46608   417.57     0.00   221.50   695.75   632.11  ...    485.62    469.37   \n",
       "\n",
       "       P2-1-H2  P2-1-H3  P2-1-H4  P2-1-H5  P2-1-H6  P2-1-H7  P2-1-H8  P2-1-H9  \n",
       "0        13.80    15.89    19.42     3.21     5.65    20.33    16.72    33.57  \n",
       "1         0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "2         0.00     0.00     8.07     0.00     0.00     0.00     2.55     0.00  \n",
       "3         0.00     4.14     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "4         0.00     0.00     0.00     0.00     0.40     0.00     1.32     0.00  \n",
       "...        ...      ...      ...      ...      ...      ...      ...      ...  \n",
       "46604     0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46605     0.00     0.00    13.15    40.04     9.40     0.00    18.05     0.00  \n",
       "46606     0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46607     0.00     2.11     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46608   290.69   135.58   196.35   388.55   357.51    83.07   220.80   408.54  \n",
       "\n",
       "[46609 rows x 385 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_1 = pd.read_csv(tpm_yang_path, sep='\\t')\n",
    "df_tpm_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clean the yang1 dataset to get the gene names, call `.shape` to check we havent filtered anything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46609, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split gene ids on _ and load into a new data frame and set the columns\n",
    "split_data = pd.DataFrame(df_tpm_1.Gene_id.str.split(\"_\", expand=True))\n",
    "split_data.columns = [\"Gene_id\", \"Gene_name\", \"Gene_name2\"]\n",
    "\n",
    "# Fill in the NA's with blank strings\n",
    "split_data[\"Gene_name\"] = split_data.Gene_name.fillna('')\n",
    "split_data[\"Gene_name2\"] = split_data.Gene_name2.fillna('')\n",
    "\n",
    "# Concatinate the strings that have split more than once back to their standard for e.g GENEID_GENENAME_SOMEMORENAME\n",
    "split_data[\"Gene_name\"] = split_data.apply(lambda x: x.Gene_name if x.Gene_name2 == '' else x.Gene_name + '_' + x.Gene_name2, axis=1)\n",
    "split_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the gene names back into the main dataset, print out dataset to check we do indeed have gene names where they were available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>...</th>\n",
       "      <th>P2-1-H11</th>\n",
       "      <th>P2-1-H12</th>\n",
       "      <th>P2-1-H2</th>\n",
       "      <th>P2-1-H3</th>\n",
       "      <th>P2-1-H4</th>\n",
       "      <th>P2-1-H5</th>\n",
       "      <th>P2-1-H6</th>\n",
       "      <th>P2-1-H7</th>\n",
       "      <th>P2-1-H8</th>\n",
       "      <th>P2-1-H9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.60</td>\n",
       "      <td>7.02</td>\n",
       "      <td>17.50</td>\n",
       "      <td>0.00</td>\n",
       "      <td>38.31</td>\n",
       "      <td>38.42</td>\n",
       "      <td>47.10</td>\n",
       "      <td>...</td>\n",
       "      <td>45.27</td>\n",
       "      <td>5.70</td>\n",
       "      <td>13.80</td>\n",
       "      <td>15.89</td>\n",
       "      <td>19.42</td>\n",
       "      <td>3.21</td>\n",
       "      <td>5.65</td>\n",
       "      <td>20.33</td>\n",
       "      <td>16.72</td>\n",
       "      <td>33.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.07</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.55</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031</td>\n",
       "      <td>H19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46604</th>\n",
       "      <td>ERCC-00164</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46605</th>\n",
       "      <td>ERCC-00165</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>42.44</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.45</td>\n",
       "      <td>9.12</td>\n",
       "      <td>...</td>\n",
       "      <td>5.41</td>\n",
       "      <td>1.76</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>13.15</td>\n",
       "      <td>40.04</td>\n",
       "      <td>9.40</td>\n",
       "      <td>0.00</td>\n",
       "      <td>18.05</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46606</th>\n",
       "      <td>ERCC-00168</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46607</th>\n",
       "      <td>ERCC-00170</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46608</th>\n",
       "      <td>ERCC-00171</td>\n",
       "      <td></td>\n",
       "      <td>0.0</td>\n",
       "      <td>143.91</td>\n",
       "      <td>1398.07</td>\n",
       "      <td>328.98</td>\n",
       "      <td>417.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>221.50</td>\n",
       "      <td>695.75</td>\n",
       "      <td>...</td>\n",
       "      <td>485.62</td>\n",
       "      <td>469.37</td>\n",
       "      <td>290.69</td>\n",
       "      <td>135.58</td>\n",
       "      <td>196.35</td>\n",
       "      <td>388.55</td>\n",
       "      <td>357.51</td>\n",
       "      <td>83.07</td>\n",
       "      <td>220.80</td>\n",
       "      <td>408.54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 386 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene_id Gene_name  P1-3-A1  P1-3-A10  P1-3-A11  P1-3-A12  \\\n",
       "0      ENSMUSG00000000001     Gnai3      0.0     17.60      7.02     17.50   \n",
       "1      ENSMUSG00000000003      Pbsn      0.0      0.00      0.00      0.00   \n",
       "2      ENSMUSG00000000028     Cdc45      0.0      0.00      0.00      0.00   \n",
       "3      ENSMUSG00000000031       H19      0.0      0.00      0.00      0.00   \n",
       "4      ENSMUSG00000000037     Scml2      0.0      0.00      0.00      0.00   \n",
       "...                   ...       ...      ...       ...       ...       ...   \n",
       "46604          ERCC-00164                0.0      0.00      0.00      0.00   \n",
       "46605          ERCC-00165                0.0      0.00     42.44      3.28   \n",
       "46606          ERCC-00168                0.0      0.00      0.00      0.00   \n",
       "46607          ERCC-00170                0.0      0.00      0.00      0.00   \n",
       "46608          ERCC-00171                0.0    143.91   1398.07    328.98   \n",
       "\n",
       "       P1-3-A2  P1-3-A3  P1-3-A4  P1-3-A5  ...  P2-1-H11  P2-1-H12  P2-1-H2  \\\n",
       "0         0.00    38.31    38.42    47.10  ...     45.27      5.70    13.80   \n",
       "1         0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "2         0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "3         0.00     0.00     0.00     0.00  ...      0.00      6.80     0.00   \n",
       "4         0.00     0.00     0.00     0.00  ...      1.47      0.00     0.00   \n",
       "...        ...      ...      ...      ...  ...       ...       ...      ...   \n",
       "46604     0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "46605     0.00     0.00     9.45     9.12  ...      5.41      1.76     0.00   \n",
       "46606     0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "46607     0.00     0.00     0.00     0.00  ...      0.00      0.00     0.00   \n",
       "46608   417.57     0.00   221.50   695.75  ...    485.62    469.37   290.69   \n",
       "\n",
       "       P2-1-H3  P2-1-H4  P2-1-H5  P2-1-H6  P2-1-H7  P2-1-H8  P2-1-H9  \n",
       "0        15.89    19.42     3.21     5.65    20.33    16.72    33.57  \n",
       "1         0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "2         0.00     8.07     0.00     0.00     0.00     2.55     0.00  \n",
       "3         4.14     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "4         0.00     0.00     0.00     0.40     0.00     1.32     0.00  \n",
       "...        ...      ...      ...      ...      ...      ...      ...  \n",
       "46604     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46605     0.00    13.15    40.04     9.40     0.00    18.05     0.00  \n",
       "46606     0.00     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46607     2.11     0.00     0.00     0.00     0.00     0.00     0.00  \n",
       "46608   135.58   196.35   388.55   357.51    83.07   220.80   408.54  \n",
       "\n",
       "[46609 rows x 386 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Insert the columns back into the main data array\n",
    "df_tpm_1[\"Gene_id\"] = split_data.Gene_id\n",
    "df_tpm_1.insert(1,\"Gene_name\", split_data.Gene_name)\n",
    "df_tpm_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create merged dataset\n",
    "\n",
    "Create merged dataset from all subsets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tpm_combined = df_tpm_1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check gene column is unique before setting is as an index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tpm_combined.Gene_id.is_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create data only set (this is easier for tensorflow to deal with)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tpm_combined_dataonly = df_tpm_combined\n",
    "df_tpm_combined_dataonly.index = df_tpm_combined['Gene_id']\n",
    "df_tpm_combined_dataonly = df_tpm_combined.drop(['Gene_id', 'Gene_name'], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create `log2(TPM+1)` dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:NumExpr defaulting to 4 threads.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P1-3-A1</th>\n",
       "      <th>P1-3-A10</th>\n",
       "      <th>P1-3-A11</th>\n",
       "      <th>P1-3-A12</th>\n",
       "      <th>P1-3-A2</th>\n",
       "      <th>P1-3-A3</th>\n",
       "      <th>P1-3-A4</th>\n",
       "      <th>P1-3-A5</th>\n",
       "      <th>P1-3-A6</th>\n",
       "      <th>P1-3-A7</th>\n",
       "      <th>...</th>\n",
       "      <th>P2-1-H11</th>\n",
       "      <th>P2-1-H12</th>\n",
       "      <th>P2-1-H2</th>\n",
       "      <th>P2-1-H3</th>\n",
       "      <th>P2-1-H4</th>\n",
       "      <th>P2-1-H5</th>\n",
       "      <th>P2-1-H6</th>\n",
       "      <th>P2-1-H7</th>\n",
       "      <th>P2-1-H8</th>\n",
       "      <th>P2-1-H9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.217231</td>\n",
       "      <td>3.003602</td>\n",
       "      <td>4.209453</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.296824</td>\n",
       "      <td>5.300856</td>\n",
       "      <td>5.587965</td>\n",
       "      <td>3.826803</td>\n",
       "      <td>3.414136</td>\n",
       "      <td>...</td>\n",
       "      <td>5.532005</td>\n",
       "      <td>2.744161</td>\n",
       "      <td>3.887525</td>\n",
       "      <td>4.078097</td>\n",
       "      <td>4.351911</td>\n",
       "      <td>2.073820</td>\n",
       "      <td>2.733354</td>\n",
       "      <td>4.414812</td>\n",
       "      <td>4.147307</td>\n",
       "      <td>5.111449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.181103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.827819</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.821838</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.963474</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.361768</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.304511</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.485427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.214125</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00164</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00165</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.440952</td>\n",
       "      <td>2.097611</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.385431</td>\n",
       "      <td>3.339137</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.599318</td>\n",
       "      <td>...</td>\n",
       "      <td>2.680324</td>\n",
       "      <td>1.464668</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.822730</td>\n",
       "      <td>5.358959</td>\n",
       "      <td>3.378512</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.251719</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00168</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00170</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.636915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC-00171</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.179013</td>\n",
       "      <td>10.450252</td>\n",
       "      <td>8.366235</td>\n",
       "      <td>8.709325</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.797662</td>\n",
       "      <td>9.444497</td>\n",
       "      <td>9.306312</td>\n",
       "      <td>6.193575</td>\n",
       "      <td>...</td>\n",
       "      <td>8.926652</td>\n",
       "      <td>8.877652</td>\n",
       "      <td>8.188292</td>\n",
       "      <td>7.093602</td>\n",
       "      <td>7.624613</td>\n",
       "      <td>8.605665</td>\n",
       "      <td>8.485870</td>\n",
       "      <td>6.393519</td>\n",
       "      <td>7.793116</td>\n",
       "      <td>8.677861</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46609 rows × 384 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    P1-3-A1  P1-3-A10   P1-3-A11  P1-3-A12   P1-3-A2  \\\n",
       "Gene_id                                                                \n",
       "ENSMUSG00000000001      0.0  4.217231   3.003602  4.209453  0.000000   \n",
       "ENSMUSG00000000003      0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000000028      0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000000031      0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000000037      0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "...                     ...       ...        ...       ...       ...   \n",
       "ERCC-00164              0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "ERCC-00165              0.0  0.000000   5.440952  2.097611  0.000000   \n",
       "ERCC-00168              0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "ERCC-00170              0.0  0.000000   0.000000  0.000000  0.000000   \n",
       "ERCC-00171              0.0  7.179013  10.450252  8.366235  8.709325   \n",
       "\n",
       "                     P1-3-A3   P1-3-A4   P1-3-A5   P1-3-A6   P1-3-A7  ...  \\\n",
       "Gene_id                                                               ...   \n",
       "ENSMUSG00000000001  5.296824  5.300856  5.587965  3.826803  3.414136  ...   \n",
       "ENSMUSG00000000003  0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSMUSG00000000028  0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ENSMUSG00000000031  0.000000  0.000000  0.000000  6.821838  0.000000  ...   \n",
       "ENSMUSG00000000037  0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "...                      ...       ...       ...       ...       ...  ...   \n",
       "ERCC-00164          0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ERCC-00165          0.000000  3.385431  3.339137  0.000000  5.599318  ...   \n",
       "ERCC-00168          0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ERCC-00170          0.000000  0.000000  0.000000  0.000000  0.000000  ...   \n",
       "ERCC-00171          0.000000  7.797662  9.444497  9.306312  6.193575  ...   \n",
       "\n",
       "                    P2-1-H11  P2-1-H12   P2-1-H2   P2-1-H3   P2-1-H4  \\\n",
       "Gene_id                                                                \n",
       "ENSMUSG00000000001  5.532005  2.744161  3.887525  4.078097  4.351911   \n",
       "ENSMUSG00000000003  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ENSMUSG00000000028  0.000000  0.000000  0.000000  0.000000  3.181103   \n",
       "ENSMUSG00000000031  0.000000  2.963474  0.000000  2.361768  0.000000   \n",
       "ENSMUSG00000000037  1.304511  0.000000  0.000000  0.000000  0.000000   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "ERCC-00164          0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ERCC-00165          2.680324  1.464668  0.000000  0.000000  3.822730   \n",
       "ERCC-00168          0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ERCC-00170          0.000000  0.000000  0.000000  1.636915  0.000000   \n",
       "ERCC-00171          8.926652  8.877652  8.188292  7.093602  7.624613   \n",
       "\n",
       "                     P2-1-H5   P2-1-H6   P2-1-H7   P2-1-H8   P2-1-H9  \n",
       "Gene_id                                                               \n",
       "ENSMUSG00000000001  2.073820  2.733354  4.414812  4.147307  5.111449  \n",
       "ENSMUSG00000000003  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "ENSMUSG00000000028  0.000000  0.000000  0.000000  1.827819  0.000000  \n",
       "ENSMUSG00000000031  0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "ENSMUSG00000000037  0.000000  0.485427  0.000000  1.214125  0.000000  \n",
       "...                      ...       ...       ...       ...       ...  \n",
       "ERCC-00164          0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "ERCC-00165          5.358959  3.378512  0.000000  4.251719  0.000000  \n",
       "ERCC-00168          0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "ERCC-00170          0.000000  0.000000  0.000000  0.000000  0.000000  \n",
       "ERCC-00171          8.605665  8.485870  6.393519  7.793116  8.677861  \n",
       "\n",
       "[46609 rows x 384 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ltpm_combined_dataonly = np.log2(df_tpm_combined_dataonly + 1)\n",
    "df_ltpm_combined_dataonly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter for genes and cells which match filtering criteria. Call `.shape` to check what was filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17139, 384)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expression_filt_mask = df_ltpm_combined_dataonly > min_ltpm_exp\n",
    "df_ltpm_combined_dataonly_genefilt = df_ltpm_combined_dataonly[df_expression_filt_mask.sum(axis=1) > min_num_cells_for_gene_exp]\n",
    "df_ltpm_combined_dataonly_genefilt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17139, 380)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ltpm_combined_dataonly_genecellfilt = df_ltpm_combined_dataonly_genefilt \\\n",
    "    .T[(df_ltpm_combined_dataonly_genefilt > min_ltpm_exp).sum(axis=0) > min_num_genes_in_cell_exp]\n",
    "df_ltpm_combined_dataonly_genecellfilt = df_ltpm_combined_dataonly_genecellfilt.T\n",
    "df_ltpm_combined_dataonly_genecellfilt.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write data to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the column and row names as a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(380, 1)\n",
      "(17139, 1)\n"
     ]
    }
   ],
   "source": [
    "df_column_names = pd.DataFrame(list(df_ltpm_combined_dataonly_genecellfilt.columns.values))\n",
    "df_row_names = pd.DataFrame(list(df_ltpm_combined_dataonly_genecellfilt.index.values))\n",
    "\n",
    "print(df_column_names.shape)\n",
    "print(df_row_names.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the data to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ltpm_combined_dataonly_genecellfilt.to_csv(data_path + '/tpm_combined.csv', index=False, header=False)\n",
    "df_column_names.to_csv(data_path + '/tpm_combined_cols.csv', index=False, header=False)\n",
    "df_row_names.to_csv(data_path + '/tpm_combined_rows.csv', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
